diff --git a/lib/marshall.js b/lib/marshall.js
index 5a7434162d97208ec34fb87bfce3c953fd90345a..86f0efd9d0c99ecb3ae4d12c46130778f813ac5b 100644
--- a/lib/marshall.js
+++ b/lib/marshall.js
@@ -1,7 +1,7 @@
 const assert = require('assert');
 
 const parseSignature = require('./signature');
-const put = require('put');
+const put = require('@nornagon/put');
 const Marshallers = require('./marshallers');
 const align = require('./align').align;
 
diff --git a/package.json b/package.json
index 2519ac38fad44569cb8a156a91af25a70c8166a5..f7d88ac6cdb00a0a7064b0e39986821816c75554 100644
--- a/package.json
+++ b/package.json
@@ -42,7 +42,7 @@
     "hexy": "^0.2.10",
     "long": "^4.0.0",
     "optimist": "^0.6.1",
-    "put": "0.0.6",
+    "@nornagon/put": "0.0.8",
     "safe-buffer": "^5.1.1",
     "xml2js": "^0.4.17"
   },
