From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: deepak1556 <hop2deep@gmail.com>
Date: Tue, 25 Aug 2026 02:51:07 +0900
Subject: 8236212: [api] Remove deprecated methods, pt.2

Ref: https://chromium-review.googlesource.com/c/v8/v8/+/8236212

diff --git a/nan_callbacks_12_inl.h b/nan_callbacks_12_inl.h
index f341221fbd55708a496c07a7fd4c7d97fa9694fb..393a07af22adb98ad40817e084b149f9ed9626e5 100644
--- a/nan_callbacks_12_inl.h
+++ b/nan_callbacks_12_inl.h
@@ -9,6 +9,10 @@
 #ifndef NAN_CALLBACKS_12_INL_H_
 #define NAN_CALLBACKS_12_INL_H_
 
+#if defined(V8_MAJOR_VERSION) && V8_MAJOR_VERSION >= 15
+# define NAN_HAS_PROPERTY_CALLBACK_INFO_BOOLEAN_SETTER 1
+#endif
+
 template<typename T>
 class ReturnValue {
   v8::ReturnValue<T> value_;
@@ -245,10 +249,22 @@ static
 void SetterCallbackWrapper(
     v8::Local<v8::Name> property
   , v8::Local<v8::Value> value
+#if defined(NAN_HAS_PROPERTY_CALLBACK_INFO_BOOLEAN_SETTER)
+  , const v8::PropertyCallbackInfo<v8::Boolean> &info) {
+  const v8::PropertyCallbackInfo<void> &void_info =
+      reinterpret_cast<const v8::PropertyCallbackInfo<void> &>(info);
+#else
   , const v8::PropertyCallbackInfo<void> &info) {
+#endif
   v8::Local<v8::Object> obj = info.Data().As<v8::Object>();
   PropertyCallbackInfo<void>
-      cbinfo(info, obj->GetInternalField(kDataIndex).As<v8::Value>());
+      cbinfo(
+#if defined(NAN_HAS_PROPERTY_CALLBACK_INFO_BOOLEAN_SETTER)
+          void_info,
+#else
+          info,
+#endif
+          obj->GetInternalField(kDataIndex).As<v8::Value>());
   SetterCallback callback = reinterpret_cast<SetterCallback>(
       reinterpret_cast<intptr_t>(
           obj->GetInternalField(kSetterIndex)
@@ -259,7 +275,11 @@ void SetterCallbackWrapper(
 typedef void (*NativeSetter)(
     v8::Local<v8::Name>
   , v8::Local<v8::Value>
+#if defined(NAN_HAS_PROPERTY_CALLBACK_INFO_BOOLEAN_SETTER)
+  , const v8::PropertyCallbackInfo<v8::Boolean> &);
+#else
   , const v8::PropertyCallbackInfo<void> &);
+#endif
 #else
 static
 void GetterCallbackWrapper(
@@ -324,10 +344,22 @@ static
 v8::Intercepted PropertySetterCallbackWrapper(
     v8::Local<v8::Name> property
   , v8::Local<v8::Value> value
+#if defined(NAN_HAS_PROPERTY_CALLBACK_INFO_BOOLEAN_SETTER)
+  , const v8::PropertyCallbackInfo<v8::Boolean> &info) {
+  const v8::PropertyCallbackInfo<void> &void_info =
+      reinterpret_cast<const v8::PropertyCallbackInfo<void> &>(info);
+#else
   , const v8::PropertyCallbackInfo<void> &info) {
+#endif
   v8::Local<v8::Object> obj = info.Data().As<v8::Object>();
   PropertyCallbackInfo<void>
-      cbinfo(info, obj->GetInternalField(kDataIndex).As<v8::Value>());
+      cbinfo(
+#if defined(NAN_HAS_PROPERTY_CALLBACK_INFO_BOOLEAN_SETTER)
+          void_info,
+#else
+          info,
+#endif
+          obj->GetInternalField(kDataIndex).As<v8::Value>());
   PropertySetterCallback callback = reinterpret_cast<PropertySetterCallback>(
       reinterpret_cast<intptr_t>(
           obj->GetInternalField(kPropertySetterIndex)
@@ -338,7 +370,11 @@ v8::Intercepted PropertySetterCallbackWrapper(
 typedef v8::Intercepted (*NativePropertySetter)(
     v8::Local<v8::Name>
   , v8::Local<v8::Value>
+#if defined(NAN_HAS_PROPERTY_CALLBACK_INFO_BOOLEAN_SETTER)
+  , const v8::PropertyCallbackInfo<v8::Boolean> &);
+#else
   , const v8::PropertyCallbackInfo<void> &);
+#endif
 
 #else
 static
@@ -577,10 +613,22 @@ static
 v8::Intercepted IndexSetterCallbackWrapper(
     uint32_t index
   , v8::Local<v8::Value> value
+#if defined(NAN_HAS_PROPERTY_CALLBACK_INFO_BOOLEAN_SETTER)
+  , const v8::PropertyCallbackInfo<v8::Boolean> &info) {
+  const v8::PropertyCallbackInfo<void> &void_info =
+      reinterpret_cast<const v8::PropertyCallbackInfo<void> &>(info);
+#else
   , const v8::PropertyCallbackInfo<void> &info) {
+#endif
   v8::Local<v8::Object> obj = info.Data().As<v8::Object>();
   PropertyCallbackInfo<void>
-      cbinfo(info, obj->GetInternalField(kDataIndex).As<v8::Value>());
+      cbinfo(
+#if defined(NAN_HAS_PROPERTY_CALLBACK_INFO_BOOLEAN_SETTER)
+          void_info,
+#else
+          info,
+#endif
+          obj->GetInternalField(kDataIndex).As<v8::Value>());
   IndexSetterCallback callback = reinterpret_cast<IndexSetterCallback>(
       reinterpret_cast<intptr_t>(
           obj->GetInternalField(kIndexPropertySetterIndex)
@@ -591,7 +639,11 @@ v8::Intercepted IndexSetterCallbackWrapper(
 typedef v8::Intercepted (*NativeIndexSetter)(
     uint32_t
   , v8::Local<v8::Value>
+#if defined(NAN_HAS_PROPERTY_CALLBACK_INFO_BOOLEAN_SETTER)
+  , const v8::PropertyCallbackInfo<v8::Boolean> &);
+#else
   , const v8::PropertyCallbackInfo<void> &);
+#endif
 
 #else
 static
