From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: deepak1556 <hop2deep@gmail.com>
Date: Tue, 25 Aug 2026 02:51:07 +0900
Subject: 8236212: [api] Remove deprecated methods, pt.2

Ref: https://chromium-review.googlesource.com/c/v8/v8/+/8236212

diff --git a/nan_callbacks_12_inl.h b/nan_callbacks_12_inl.h
index 8ec91f84d7ca957ca7b366619e3e7a2ed68709f3..7c0b08298c565486d5669b288271caf8573657bd 100644
--- a/nan_callbacks_12_inl.h
+++ b/nan_callbacks_12_inl.h
@@ -16,6 +16,10 @@
 # define NAN_HAS_PROPERTY_CALLBACK_INFO_HOLDER_V2 1
 #endif
 
+#if defined(V8_MAJOR_VERSION) && V8_MAJOR_VERSION >= 15
+# define NAN_HAS_PROPERTY_CALLBACK_INFO_BOOLEAN_SETTER 1
+#endif
+
 template<typename T>
 class ReturnValue {
   v8::ReturnValue<T> value_;
@@ -253,10 +257,22 @@ static
 void SetterCallbackWrapper(
     v8::Local<v8::Name> property
   , v8::Local<v8::Value> value
+#if defined(NAN_HAS_PROPERTY_CALLBACK_INFO_BOOLEAN_SETTER)
+  , const v8::PropertyCallbackInfo<v8::Boolean> &info) {
+  const v8::PropertyCallbackInfo<void> &void_info =
+      reinterpret_cast<const v8::PropertyCallbackInfo<void> &>(info);
+#else
   , const v8::PropertyCallbackInfo<void> &info) {
+#endif
   v8::Local<v8::Object> obj = info.Data().As<v8::Object>();
   PropertyCallbackInfo<void>
-      cbinfo(info, obj->GetInternalField(kDataIndex).As<v8::Value>());
+      cbinfo(
+#if defined(NAN_HAS_PROPERTY_CALLBACK_INFO_BOOLEAN_SETTER)
+          void_info,
+#else
+          info,
+#endif
+          obj->GetInternalField(kDataIndex).As<v8::Value>());
   SetterCallback callback = reinterpret_cast<SetterCallback>(
       reinterpret_cast<intptr_t>(
           Nan::GetExternalValue(obj->GetInternalField(kSetterIndex)
@@ -267,7 +283,11 @@ void SetterCallbackWrapper(
 typedef void (*NativeSetter)(
     v8::Local<v8::Name>
   , v8::Local<v8::Value>
+#if defined(NAN_HAS_PROPERTY_CALLBACK_INFO_BOOLEAN_SETTER)
+  , const v8::PropertyCallbackInfo<v8::Boolean> &);
+#else
   , const v8::PropertyCallbackInfo<void> &);
+#endif
 #else
 static
 void GetterCallbackWrapper(
@@ -332,10 +352,22 @@ static
 v8::Intercepted PropertySetterCallbackWrapper(
     v8::Local<v8::Name> property
   , v8::Local<v8::Value> value
+#if defined(NAN_HAS_PROPERTY_CALLBACK_INFO_BOOLEAN_SETTER)
+  , const v8::PropertyCallbackInfo<v8::Boolean> &info) {
+  const v8::PropertyCallbackInfo<void> &void_info =
+      reinterpret_cast<const v8::PropertyCallbackInfo<void> &>(info);
+#else
   , const v8::PropertyCallbackInfo<void> &info) {
+#endif
   v8::Local<v8::Object> obj = info.Data().As<v8::Object>();
   PropertyCallbackInfo<void>
-      cbinfo(info, obj->GetInternalField(kDataIndex).As<v8::Value>());
+      cbinfo(
+#if defined(NAN_HAS_PROPERTY_CALLBACK_INFO_BOOLEAN_SETTER)
+          void_info,
+#else
+          info,
+#endif
+          obj->GetInternalField(kDataIndex).As<v8::Value>());
   PropertySetterCallback callback = reinterpret_cast<PropertySetterCallback>(
       reinterpret_cast<intptr_t>(
           Nan::GetExternalValue(obj->GetInternalField(kPropertySetterIndex)
@@ -346,7 +378,11 @@ v8::Intercepted PropertySetterCallbackWrapper(
 typedef v8::Intercepted (*NativePropertySetter)(
     v8::Local<v8::Name>
   , v8::Local<v8::Value>
+#if defined(NAN_HAS_PROPERTY_CALLBACK_INFO_BOOLEAN_SETTER)
+  , const v8::PropertyCallbackInfo<v8::Boolean> &);
+#else
   , const v8::PropertyCallbackInfo<void> &);
+#endif
 
 #else
 static
@@ -585,10 +621,22 @@ static
 v8::Intercepted IndexSetterCallbackWrapper(
     uint32_t index
   , v8::Local<v8::Value> value
+#if defined(NAN_HAS_PROPERTY_CALLBACK_INFO_BOOLEAN_SETTER)
+  , const v8::PropertyCallbackInfo<v8::Boolean> &info) {
+  const v8::PropertyCallbackInfo<void> &void_info =
+      reinterpret_cast<const v8::PropertyCallbackInfo<void> &>(info);
+#else
   , const v8::PropertyCallbackInfo<void> &info) {
+#endif
   v8::Local<v8::Object> obj = info.Data().As<v8::Object>();
   PropertyCallbackInfo<void>
-      cbinfo(info, obj->GetInternalField(kDataIndex).As<v8::Value>());
+      cbinfo(
+#if defined(NAN_HAS_PROPERTY_CALLBACK_INFO_BOOLEAN_SETTER)
+          void_info,
+#else
+          info,
+#endif
+          obj->GetInternalField(kDataIndex).As<v8::Value>());
   IndexSetterCallback callback = reinterpret_cast<IndexSetterCallback>(
       reinterpret_cast<intptr_t>(
           Nan::GetExternalValue(obj->GetInternalField(kIndexPropertySetterIndex)
@@ -599,7 +647,11 @@ v8::Intercepted IndexSetterCallbackWrapper(
 typedef v8::Intercepted (*NativeIndexSetter)(
     uint32_t
   , v8::Local<v8::Value>
+#if defined(NAN_HAS_PROPERTY_CALLBACK_INFO_BOOLEAN_SETTER)
+  , const v8::PropertyCallbackInfo<v8::Boolean> &);
+#else
   , const v8::PropertyCallbackInfo<void> &);
+#endif
 
 #else
 static
