diff --git a/node_modules/@octokit/request/dist-src/fetch-wrapper.js b/node_modules/@octokit/request/dist-src/fetch-wrapper.js
index 79653c4..ac608fd 100644
--- a/node_modules/@octokit/request/dist-src/fetch-wrapper.js
+++ b/node_modules/@octokit/request/dist-src/fetch-wrapper.js
@@ -19,6 +19,9 @@ export default function fetchWrapper(requestOptions) {
         body: requestOptions.body,
         headers: requestOptions.headers,
         redirect: requestOptions.redirect,
+        // duplex must be set if request.body is ReadableStream or Async Iterables.
+        // See https://fetch.spec.whatwg.org/#dom-requestinit-duplex.
+        ...(requestOptions.body && { duplex: "half" }),
     }, 
     // `requestOptions.request.agent` type is incompatible
     // see https://github.com/octokit/types.ts/pull/264
