b767cde1e7bb94b3fea13f2d7e10cd90e6813d6c
diff --git a/src/runtime/runtime-intl.cc b/src/runtime/runtime-intl.cc
index c359a1e5a1..198610a652 100644
--- a/src/runtime/runtime-intl.cc
+++ b/src/runtime/runtime-intl.cc
@@ -620,8 +620,7 @@ RUNTIME_FUNCTION(Runtime_PluralRulesSelect) {
   icu::UnicodeString result = plural_rules->select(rounded);
   return *isolate->factory()
               ->NewStringFromTwoByte(Vector<const uint16_t>(
-                  reinterpret_cast<const uint16_t*>(
-                      icu::toUCharPtr(result.getBuffer())),
+                  reinterpret_cast<const uint16_t*>(result.getBuffer()),
                   result.length()))
               .ToHandleChecked();
 }
