From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ales Pergl <alpergl@microsoft.com>
Date: Mon, 22 Oct 2018 10:47:12 -0700
Subject: build_gn.patch


diff --git a/BUILD.gn b/BUILD.gn
index 38783c0dace2fe896f95d1dcb8aefef16d3a5a7d..20e633a3cded94be71b3cfbaf0c5a3d1d57f6158 100644
--- a/BUILD.gn
+++ b/BUILD.gn
@@ -262,7 +262,7 @@ config("internal_config") {
     ":v8_header_features",
   ]
 
-  if (is_component_build) {
+  if (is_component_build || is_electron_build) {
     defines += [ "BUILDING_V8_SHARED" ]
   }
 }
@@ -3684,6 +3684,8 @@ if (v8_use_snapshot && current_toolchain == v8_snapshot_toolchain) {
 
     configs = [ ":internal_config" ]
 
+    configs += [ "//electron/build/config:build_time_executable" ]
+
     deps = [
       ":v8_base_without_compiler",
       ":v8_compiler_for_mksnapshot",
