From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aleksei Kuzmin <alkuzmin@microsoft.com>
Date: Mon, 22 Oct 2018 10:47:13 -0700
Subject: build-torque-with-x64-toolchain-on-arm.patch

torque binary has to be run during the build.

diff --git a/BUILD.gn b/BUILD.gn
index c1b08958b500da1910a8067198cdec7650534f20..78e672e18eff27a3777bd0c0adc94e88a6eb126a 100644
--- a/BUILD.gn
+++ b/BUILD.gn
@@ -914,7 +914,8 @@ action("run_torque") {
   # is the target toolchain and, hence, can't be used.
   v8_torque_toolchain = v8_snapshot_toolchain
   if (host_cpu == "x64" &&
-      (v8_current_cpu == "mips" || v8_current_cpu == "mips64")) {
+      (v8_current_cpu == "mips" || v8_current_cpu == "mips64" ||
+       v8_current_cpu == "arm" || v8_current_cpu == "arm64")) {
     v8_torque_toolchain = "//build/toolchain/linux:clang_x64"
   }
 
@@ -3256,7 +3257,7 @@ if (v8_use_snapshot && current_toolchain == v8_snapshot_toolchain) {
   }
 }
 
-if (current_toolchain == v8_snapshot_toolchain) {
+if (current_toolchain == current_toolchain) {
   v8_executable("torque") {
     visibility = [ ":*" ]  # Only targets in this file can depend on this.
 
