From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ales Pergl <alpergl@microsoft.com>
Date: Mon, 22 Oct 2018 10:47:12 -0700
Subject: build_gn.patch


diff --git a/BUILD.gn b/BUILD.gn
index cc9b7a906a111451eff86e8f78571617256136f1..139a777e304ab0449693664feb248a714995aa26 100644
--- a/BUILD.gn
+++ b/BUILD.gn
@@ -236,7 +236,7 @@ config("internal_config") {
 
   configs = [ "//build/config/compiler:wexit_time_destructors" ]
 
-  if (is_component_build) {
+  if (is_component_build || is_electron_build) {
     defines = [ "BUILDING_V8_SHARED" ]
   }
 }
@@ -3386,6 +3386,8 @@ if (v8_use_snapshot && current_toolchain == v8_snapshot_toolchain) {
 
     configs = [ ":internal_config" ]
 
+    configs += [ "//electron/build/config:build_time_executable" ]
+
     deps = [
       ":v8_base",
       ":v8_init",
