From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ales Pergl <alpergl@microsoft.com>
Date: Mon, 22 Oct 2018 10:47:12 -0700
Subject: build_gn.patch


diff --git a/BUILD.gn b/BUILD.gn
index 04a2c136f409a9ff2bc0b01a48acb5f7c81ad3e3..83f570c9493ca388eb5718d414bacd4b8cd37c6d 100644
--- a/BUILD.gn
+++ b/BUILD.gn
@@ -262,7 +262,7 @@ config("internal_config") {
     ":v8_header_features",
   ]
 
-  if (is_component_build) {
+  if (is_component_build || is_electron_build) {
     defines += [ "BUILDING_V8_SHARED" ]
   }
 }
@@ -3580,6 +3580,8 @@ if (v8_use_snapshot && current_toolchain == v8_snapshot_toolchain) {
 
     configs = [ ":internal_config" ]
 
+    configs += [ "//electron/build/config:build_time_executable" ]
+
     deps = [
       ":v8_base",
       ":v8_init",
