diff --git a/BUILD.gn b/BUILD.gn
index 456a318c1c..283bce15f0 100644
--- a/BUILD.gn
+++ b/BUILD.gn
@@ -195,7 +195,7 @@ config("internal_config") {
     "$target_gen_dir",
   ]
 
-  if (is_component_build) {
+  if (is_component_build || is_electron_build) {
     defines = [ "BUILDING_V8_SHARED" ]
   }
 }
@@ -2674,6 +2674,8 @@ if (current_toolchain == v8_snapshot_toolchain) {
 
     configs = [ ":internal_config" ]
 
+    configs += [ "//electron/build/config:build_time_executable" ]
+
     deps = [
       ":v8_base",
       ":v8_init",
