From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: deepak1556 <hop2deep@gmail.com>
Date: Fri, 4 Jan 2019 15:50:24 +0530
Subject: build-torque-with-x64-toolchain-on-arm.patch

torque binary has to be run during the build.

diff --git a/BUILD.gn b/BUILD.gn
index b70c09aa34aa7547e7d26d9e35795904a17c092f..ff8d1fbeabfd9730df4488981397c810af099d7e 100644
--- a/BUILD.gn
+++ b/BUILD.gn
@@ -184,7 +184,9 @@ declare_args() {
 
 v8_generator_toolchain = v8_snapshot_toolchain
 if (host_cpu == "x64" &&
-    (v8_current_cpu == "mips" || v8_current_cpu == "mips64")) {
+    (v8_current_cpu == "mips" || v8_current_cpu == "mips64" ||
+     v8_snapshot_toolchain == "//build/toolchain/linux:clang_arm" ||
+     v8_snapshot_toolchain == "//build/toolchain/linux:clang_arm64")) {
   v8_generator_toolchain = "//build/toolchain/linux:clang_x64"
 }
 
@@ -3402,7 +3404,7 @@ if (v8_use_snapshot && current_toolchain == v8_snapshot_toolchain) {
   }
 }
 
-if (current_toolchain == v8_snapshot_toolchain) {
+if (current_toolchain == v8_generator_toolchain) {
   v8_executable("torque") {
     visibility = [ ":*" ]  # Only targets in this file can depend on this.
 
