diff --git a/BUILD.gn b/BUILD.gn
index 456a318..af803ff 100644
--- a/BUILD.gn
+++ b/BUILD.gn
@@ -870,7 +870,8 @@ action("run_torque") {
   # is the target toolchain and, hence, can't be used.
   v8_torque_toolchain = v8_snapshot_toolchain
   if (host_cpu == "x64" &&
-      (v8_current_cpu == "mips" || v8_current_cpu == "mips64")) {
+      (v8_current_cpu == "mips" || v8_current_cpu == "mips64" ||
+       v8_current_cpu == "arm" || v8_current_cpu == "arm64")) {
     v8_torque_toolchain = "//build/toolchain/linux:clang_x64"
   }
 
@@ -3079,7 +3080,7 @@ if (v8_use_snapshot && current_toolchain == v8_snapshot_toolchain) {
   }
 }
 
-if (current_toolchain == v8_snapshot_toolchain) {
+if (current_toolchain == current_toolchain) {
   v8_executable("torque") {
     visibility = [ ":*" ]  # Only targets in this file can depend on this.
 
