From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ales Pergl <alpergl@microsoft.com>
Date: Mon, 22 Oct 2018 10:47:12 -0700
Subject: dcheck.patch


diff --git a/src/api.cc b/src/api.cc
index b14e53bc93102157841316d1fdbdb524a1a89c73..63a0ed79fb8dbc78db0ff9578ef1bb5cfc1663f5 100644
--- a/src/api.cc
+++ b/src/api.cc
@@ -8628,7 +8628,7 @@ void Isolate::SetPromiseRejectCallback(PromiseRejectCallback callback) {
 
 
 void Isolate::RunMicrotasks() {
-  DCHECK_NE(MicrotasksPolicy::kScoped, GetMicrotasksPolicy());
+  // DCHECK_NE(MicrotasksPolicy::kScoped, GetMicrotasksPolicy());
   i::Isolate* isolate = reinterpret_cast<i::Isolate*>(this);
   isolate->default_microtask_queue()->RunMicrotasks(isolate);
 }
diff --git a/src/heap/heap.cc b/src/heap/heap.cc
index 0462e9ee245cbfec985ae1a2ffe4f6699531e3aa..3c312c97f152846980073ab4187ac47fa388b46f 100644
--- a/src/heap/heap.cc
+++ b/src/heap/heap.cc
@@ -4848,9 +4848,9 @@ void Heap::TearDown() {
 void Heap::AddGCPrologueCallback(v8::Isolate::GCCallbackWithData callback,
                                  GCType gc_type, void* data) {
   DCHECK_NOT_NULL(callback);
-  DCHECK(gc_prologue_callbacks_.end() ==
-         std::find(gc_prologue_callbacks_.begin(), gc_prologue_callbacks_.end(),
-                   GCCallbackTuple(callback, gc_type, data)));
+//  DCHECK(gc_prologue_callbacks_.end() ==
+//         std::find(gc_prologue_callbacks_.begin(), gc_prologue_callbacks_.end(),
+//                   GCCallbackTuple(callback, gc_type, data)));
   gc_prologue_callbacks_.emplace_back(callback, gc_type, data);
 }
 
