From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ales Pergl <alpergl@microsoft.com>
Date: Mon, 22 Oct 2018 10:47:12 -0700
Subject: dcheck.patch


diff --git a/src/api.cc b/src/api.cc
index 679f7db6724976ae960182e838133d672c17f312..bd7cfa48aa219263551b7a9e523ae82324cac492 100644
--- a/src/api.cc
+++ b/src/api.cc
@@ -8592,7 +8592,7 @@ void Isolate::SetPromiseRejectCallback(PromiseRejectCallback callback) {
 
 
 void Isolate::RunMicrotasks() {
-  DCHECK_NE(MicrotasksPolicy::kScoped, GetMicrotasksPolicy());
+  // DCHECK_NE(MicrotasksPolicy::kScoped, GetMicrotasksPolicy());
   i::Isolate* isolate = reinterpret_cast<i::Isolate*>(this);
   isolate->default_microtask_queue()->RunMicrotasks(isolate);
 }
diff --git a/src/heap/heap.cc b/src/heap/heap.cc
index 4e1b1b8429b42f9d6b20faf60e45911656699021..c734ef5cb206a2d74c64f385ec45cb45fe38cef1 100644
--- a/src/heap/heap.cc
+++ b/src/heap/heap.cc
@@ -4944,9 +4944,9 @@ void Heap::TearDown() {
 void Heap::AddGCPrologueCallback(v8::Isolate::GCCallbackWithData callback,
                                  GCType gc_type, void* data) {
   DCHECK_NOT_NULL(callback);
-  DCHECK(gc_prologue_callbacks_.end() ==
-         std::find(gc_prologue_callbacks_.begin(), gc_prologue_callbacks_.end(),
-                   GCCallbackTuple(callback, gc_type, data)));
+//  DCHECK(gc_prologue_callbacks_.end() ==
+//         std::find(gc_prologue_callbacks_.begin(), gc_prologue_callbacks_.end(),
+//                   GCCallbackTuple(callback, gc_type, data)));
   gc_prologue_callbacks_.emplace_back(callback, gc_type, data);
 }
 
