From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samuel Attard <sattard@slack-corp.com>
Date: Thu, 27 Feb 2020 11:47:31 -0800
Subject: Fix bug in receiver maps inference

Refs: https://chromium-review.googlesource.com/c/v8/v8/+/2062404

diff --git a/src/compiler/node-properties.cc b/src/compiler/node-properties.cc
index d6528c553a149cb7bcf1543b0e0c3b13d30622b2..92226565ea732d197e582f31d53d28a86236ed2f 100644
--- a/src/compiler/node-properties.cc
+++ b/src/compiler/node-properties.cc
@@ -444,6 +444,7 @@ NodeProperties::InferReceiverMapsResult NodeProperties::InferReceiverMaps(
           // We reached the allocation of the {receiver}.
           return kNoReceiverMaps;
         }
+        result = kUnreliableReceiverMaps;  // JSCreate can have side-effect.
         break;
       }
       case IrOpcode::kJSCreatePromise: {
