From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jeremy Apthorp <jeremya@chromium.org>
Date: Mon, 15 Apr 2019 18:11:56 -0700
Subject: fixup! build_gn.patch


diff --git a/BUILD.gn b/BUILD.gn
index 83af921ee204545d215b9f0116b832e199f11e02..b4350c88822023f5c52c081746de034401d0c47c 100644
--- a/BUILD.gn
+++ b/BUILD.gn
@@ -3655,7 +3655,7 @@ if (current_toolchain == v8_generator_toolchain) {
       "src/interpreter/bytecodes.h",
     ]
 
-    configs = [ ":internal_config" ]
+    configs = [ ":internal_config_base" ]
 
     deps = [
       ":v8_libbase",
