diff --git a/src/ostreams.h b/src/ostreams.h
index c6b64a1cd9..236a73d018 100644
--- a/src/ostreams.h
+++ b/src/ostreams.h
@@ -32,7 +32,7 @@ class OFStreamBase : public std::streambuf {
 };
 
 // An output stream writing to a file.
-class V8_EXPORT_PRIVATE OFStream : public std::ostream {
+class V8_EXPORT_PRIVATE OFStream : public NON_EXPORTED_BASE(std::ostream) {
  public:
   explicit OFStream(FILE* f);
   virtual ~OFStream();
