From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Cheng Zhao <zcbenz@gmail.com>
Date: Mon, 22 Oct 2018 10:47:12 -0700
Subject: ostreams.patch


diff --git a/src/ostreams.h b/src/ostreams.h
index 189f5384b99438f3dad1d77a0836ceb45f6ed938..83a5930d4bd690dde01749bcd88acb11d75aa0cb 100644
--- a/src/ostreams.h
+++ b/src/ostreams.h
@@ -32,7 +32,7 @@ class V8_EXPORT_PRIVATE OFStreamBase : public std::streambuf {
 };
 
 // An output stream writing to a file.
-class V8_EXPORT_PRIVATE OFStream : public std::ostream {
+class V8_EXPORT_PRIVATE OFStream : public NON_EXPORTED_BASE(std::ostream) {
  public:
   explicit OFStream(FILE* f);
   ~OFStream() override = default;
