diff --git a/BUILD.gn b/BUILD.gn
index 76914cd7..a44d7896 100644
--- a/BUILD.gn
+++ b/BUILD.gn
@@ -34,7 +34,7 @@ config("icu_config") {
     "U_ENABLE_DYLOAD=0",
   ]
 
-  if (!is_component_build) {
+  if (!is_component_build && !(is_electron_build && !is_electron_gn_build)) {
     defines += [ "U_STATIC_IMPLEMENTATION" ]
   }
 
@@ -537,6 +537,24 @@ component("icui18n") {
     ":icuuc",
   ]
 
+  # In Electron all of ICU is in one binary, hence we must build with
+  # U_COMBINED_IMPLEMENTATION defined.
+  # Also, for the "static_library" configuration, keep icui18n as
+  # a shared library so that other Chromium targets build cleanly.
+  if (is_electron_build && !is_electron_gn_build) {
+    defines += [ "U_COMBINED_IMPLEMENTATION" ]
+
+    if (!is_component_build) {
+      static_component_type = "shared_library"
+
+      # Because GN already applied the wrong defaults (see the
+      # set_defaults("component") statement in BUILDCONFIG.gn) we must reset
+      # the configs and assign those for shared libraries.
+      configs = []
+      configs = default_shared_library_configs
+    }
+  }
+
   # ICU uses RTTI, replace the default "no rtti" config.
   configs -= [
     "//build/config/compiler:no_rtti",  # ICU uses RTTI.
@@ -945,6 +963,24 @@ component("icuuc") {
     ":icudata",
   ]
 
+  # In Electron all of ICU is in one binary, hence we must build with
+  # U_COMBINED_IMPLEMENTATION defined.
+  # Also, for the "static_library" configuration, keep icuuc as
+  # a shared library so that other Chromium targets build cleanly.
+  if (is_electron_build && !is_electron_gn_build) {
+    defines += [ "U_COMBINED_IMPLEMENTATION" ]
+
+    if (!is_component_build) {
+      static_component_type = "shared_library"
+
+      # Because GN already applied the wrong defaults (see the
+      # set_defaults("component") statement in BUILDCONFIG.gn) we must reset
+      # the configs and assign those for shared libraries.
+      configs = []
+      configs = default_shared_library_configs
+    }
+  }
+
   configs -= [
     "//build/config/compiler:no_rtti",  # ICU uses RTTI.
     "//build/config/compiler:chromium_code",
