From dcd2071f20aa79f2eb1959cd5f040a0e4a0c058b Mon Sep 17 00:00:00 2001
From: Wang Qing <wangqing-hf@loongson.cn>
Date: Sat, 30 Sep 2017 10:21:25 +0800
Subject: [PATCH] Fix build_ffmpeg with cross-prefix mips64el-redhat-linux-.

---
 chromium/scripts/build_ffmpeg.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/chromium/scripts/build_ffmpeg.py b/chromium/scripts/build_ffmpeg.py
index 0dbd9eb..79b5af2 100755
--- a/chromium/scripts/build_ffmpeg.py
+++ b/chromium/scripts/build_ffmpeg.py
@@ -516,7 +516,7 @@ def main(argv):
         else:
           configure_flags['Common'].extend([
             '--enable-cross-compile',
-            '--cross-prefix=mips64el-linux-gnuabi64-',
+            '--cross-prefix=mips64el-loongson-linux-',
             '--target-os=linux',
             '--arch=mips',
             '--extra-cflags=-mips64r2',
