From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: John Kleinschmidt <jkleinsc@electronjs.org>
Date: Wed, 11 Jun 2025 14:40:51 -0400
Subject: chore: debug x_error_trap dcheck

Try to figure out why this dcheck is firing.

diff --git a/modules/desktop_capture/linux/x11/x_error_trap.cc b/modules/desktop_capture/linux/x11/x_error_trap.cc
index 9199a99177c105cdf70957c25af187fcb42cb06b..d7dfeaa062642faf245aaeb8ae4118fd2d077d4a 100644
--- a/modules/desktop_capture/linux/x11/x_error_trap.cc
+++ b/modules/desktop_capture/linux/x11/x_error_trap.cc
@@ -8,6 +8,7 @@
  *  be found in the AUTHORS file in the root of the source tree.
  */
 
+#include "base/logging.h"
 #include "modules/desktop_capture/linux/x11/x_error_trap.h"
 
 #include <X11/Xlib.h>
@@ -31,7 +32,13 @@ Mutex* AcquireMutex() {
 }
 
 int XServerErrorHandler(Display* display, XErrorEvent* error_event) {
-  RTC_DCHECK_EQ(display, g_display_for_error_handler.load());
+  LOG(INFO) << "X server error: "
+            << "display=" << display << ", "
+            << "g_display_for_error_handler=" << g_display_for_error_handler.load() << ", "
+            << "error_code=" << error_event->error_code << ", "
+            << "request_code=" << error_event->request_code << ", "
+            << "minor_code=" << error_event->minor_code;
+  //RTC_DCHECK_EQ(display, g_display_for_error_handler.load());
   g_last_xserver_error_code = error_event->error_code;
   return 0;
 }
