From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jeremy Apthorp <nornagon@nornagon.net>
Date: Wed, 12 Jun 2019 10:35:22 -0700
Subject: Fix link error on arm platform


diff --git a/third_party/llvm-7.0/BUILD.gn b/third_party/llvm-7.0/BUILD.gn
index 6fc86074e7ae5c777845bb6523d88bbad539aec3..3b3afb820139fc7f330019e80861b9b98899b98c 100644
--- a/third_party/llvm-7.0/BUILD.gn
+++ b/third_party/llvm-7.0/BUILD.gn
@@ -140,6 +140,7 @@ swiftshader_source_set("swiftshader_llvm") {
     "llvm/lib/Analysis/LazyBranchProbabilityInfo.cpp",
     "llvm/lib/Analysis/LazyValueInfo.cpp",
     "llvm/lib/Analysis/Loads.cpp",
+    "llvm/lib/Analysis/LoopAccessAnalysis.cpp",
     "llvm/lib/Analysis/LoopAnalysisManager.cpp",
     "llvm/lib/Analysis/LoopInfo.cpp",
     "llvm/lib/Analysis/LoopPass.cpp",
@@ -1015,6 +1016,7 @@ swiftshader_source_set("swiftshader_llvm_transforms") {
     "llvm/lib/Transforms/Scalar/LICM.cpp",
     "llvm/lib/Transforms/Scalar/LoopDataPrefetch.cpp",
     "llvm/lib/Transforms/Scalar/LoopStrengthReduce.cpp",
+    "llvm/lib/Transforms/Scalar/LowerAtomic.cpp",
     "llvm/lib/Transforms/Scalar/MergeICmps.cpp",
     "llvm/lib/Transforms/Scalar/PartiallyInlineLibCalls.cpp",
     "llvm/lib/Transforms/Scalar/Reassociate.cpp",
