From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Cheng Zhao <zcbenz@gmail.com>
Date: Wed, 22 Nov 2023 17:21:29 +0900
Subject: build: expose mksnapshot to embedders

Backport: https://chromium-review.googlesource.com/c/v8/v8/+/5052522

diff --git a/BUILD.gn b/BUILD.gn
index 17a0e5dd1a693b0f461396d9dcfb9ea812885472..206f1084252a264b060239e7218331d88a858764 100644
--- a/BUILD.gn
+++ b/BUILD.gn
@@ -6913,8 +6913,6 @@ if (current_toolchain == v8_generator_toolchain) {
 
 if (current_toolchain == v8_snapshot_toolchain) {
   v8_executable("mksnapshot") {
-    visibility = [ ":*" ]  # Only targets in this file can depend on this.
-
     sources = [
       "src/snapshot/embedded/embedded-empty.cc",
       "src/snapshot/embedded/embedded-file-writer.cc",
