From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ales Pergl <alpergl@microsoft.com>
Date: Mon, 22 Oct 2018 10:47:12 -0700
Subject: dcheck.patch

https://github.com/auchenberg/volkswagen

diff --git a/src/api/api.cc b/src/api/api.cc
index 8e8d5130b8166ea5d2a1c2ddbc26540a03d12136..0564520fc28cb4c36bd5bb1c4b98225b5c19726e 100644
--- a/src/api/api.cc
+++ b/src/api/api.cc
@@ -9883,7 +9883,7 @@ void Isolate::SetPromiseRejectCallback(PromiseRejectCallback callback) {
 }
 
 void Isolate::PerformMicrotaskCheckpoint() {
-  DCHECK_NE(MicrotasksPolicy::kScoped, GetMicrotasksPolicy());
+  // DCHECK_NE(MicrotasksPolicy::kScoped, GetMicrotasksPolicy());
   i::Isolate* i_isolate = reinterpret_cast<i::Isolate*>(this);
   i_isolate->default_microtask_queue()->PerformCheckpoint(this);
 }
