From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ales Pergl <alpergl@microsoft.com>
Date: Mon, 22 Oct 2018 10:47:12 -0700
Subject: dcheck.patch

https://github.com/auchenberg/volkswagen

diff --git a/src/api/api.cc b/src/api/api.cc
index bf0e1b2a3e661fdb38a91e1a62a12a6da65cab98..6009c0526d5b713f5899d99e726b832ae2735e8d 100644
--- a/src/api/api.cc
+++ b/src/api/api.cc
@@ -9116,7 +9116,7 @@ void Isolate::SetPromiseRejectCallback(PromiseRejectCallback callback) {
 }
 
 void Isolate::PerformMicrotaskCheckpoint() {
-  DCHECK_NE(MicrotasksPolicy::kScoped, GetMicrotasksPolicy());
+  // DCHECK_NE(MicrotasksPolicy::kScoped, GetMicrotasksPolicy());
   i::Isolate* i_isolate = reinterpret_cast<i::Isolate*>(this);
   i_isolate->default_microtask_queue()->PerformCheckpoint(this);
 }
diff --git a/src/heap/heap.cc b/src/heap/heap.cc
index 78462e99f5724e0a8a6ffcc6b1518218a7f76bbf..d74fde9829ecd132f02fe1140e2e0a2c066402c8 100644
--- a/src/heap/heap.cc
+++ b/src/heap/heap.cc
@@ -6210,9 +6210,9 @@ void Heap::TearDown() {
 void Heap::AddGCPrologueCallback(v8::Isolate::GCCallbackWithData callback,
                                  GCType gc_type, void* data) {
   DCHECK_NOT_NULL(callback);
-  DCHECK(gc_prologue_callbacks_.end() ==
-         std::find(gc_prologue_callbacks_.begin(), gc_prologue_callbacks_.end(),
-                   GCCallbackTuple(callback, gc_type, data)));
+  // DCHECK(gc_prologue_callbacks_.end() ==
+  //        std::find(gc_prologue_callbacks_.begin(), gc_prologue_callbacks_.end(),
+  //                  GCCallbackTuple(callback, gc_type, data)));
   gc_prologue_callbacks_.emplace_back(callback, gc_type, data);
 }
 
