From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ales Pergl <alpergl@microsoft.com>
Date: Mon, 22 Oct 2018 10:47:12 -0700
Subject: dcheck.patch

https://github.com/auchenberg/volkswagen

diff --git a/src/api/api.cc b/src/api/api.cc
index 0e3370e84503dcdece061dd073b337f774b2d2da..7b8eb622c50a081c8fc3a365f1f4c7e017084e06 100644
--- a/src/api/api.cc
+++ b/src/api/api.cc
@@ -9068,7 +9068,7 @@ void Isolate::SetPromiseRejectCallback(PromiseRejectCallback callback) {
 }
 
 void Isolate::PerformMicrotaskCheckpoint() {
-  DCHECK_NE(MicrotasksPolicy::kScoped, GetMicrotasksPolicy());
+  //DCHECK_NE(MicrotasksPolicy::kScoped, GetMicrotasksPolicy());
   i::Isolate* isolate = reinterpret_cast<i::Isolate*>(this);
   isolate->default_microtask_queue()->PerformCheckpoint(this);
 }
diff --git a/src/heap/heap.cc b/src/heap/heap.cc
index aefbbd9b31e0db74c33858842ff4856eb9962706..7d348738a40d6e27b55b18aea7e8fea553834fd7 100644
--- a/src/heap/heap.cc
+++ b/src/heap/heap.cc
@@ -6111,9 +6111,9 @@ void Heap::TearDown() {
 void Heap::AddGCPrologueCallback(v8::Isolate::GCCallbackWithData callback,
                                  GCType gc_type, void* data) {
   DCHECK_NOT_NULL(callback);
-  DCHECK(gc_prologue_callbacks_.end() ==
-         std::find(gc_prologue_callbacks_.begin(), gc_prologue_callbacks_.end(),
-                   GCCallbackTuple(callback, gc_type, data)));
+//  DCHECK(gc_prologue_callbacks_.end() ==
+//         std::find(gc_prologue_callbacks_.begin(), gc_prologue_callbacks_.end(),
+//                   GCCallbackTuple(callback, gc_type, data)));
   gc_prologue_callbacks_.emplace_back(callback, gc_type, data);
 }
 
