From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ales Pergl <alpergl@microsoft.com>
Date: Mon, 22 Oct 2018 10:47:12 -0700
Subject: dcheck.patch

https://github.com/auchenberg/volkswagen

diff --git a/src/api/api.cc b/src/api/api.cc
index 81dfbb0abeea4c742afa179f174560aac56699b8..464d51bc4c8cc52c94557dec0db9e881578d849a 100644
--- a/src/api/api.cc
+++ b/src/api/api.cc
@@ -9885,7 +9885,7 @@ void Isolate::SetPromiseRejectCallback(PromiseRejectCallback callback) {
 }
 
 void Isolate::PerformMicrotaskCheckpoint() {
-  DCHECK_NE(MicrotasksPolicy::kScoped, GetMicrotasksPolicy());
+  // DCHECK_NE(MicrotasksPolicy::kScoped, GetMicrotasksPolicy());
   i::Isolate* i_isolate = reinterpret_cast<i::Isolate*>(this);
   i_isolate->default_microtask_queue()->PerformCheckpoint(this);
 }
