From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ales Pergl <alpergl@microsoft.com>
Date: Mon, 22 Oct 2018 10:47:12 -0700
Subject: dcheck.patch

https://github.com/auchenberg/volkswagen

diff --git a/src/api/api.cc b/src/api/api.cc
index 608da3369fabaf08db5ccb0b92e2ff8c5db9957d..1b2131282a1e6a60ab02a9d58ef0320825b6e60b 100644
--- a/src/api/api.cc
+++ b/src/api/api.cc
@@ -9244,7 +9244,7 @@ void Isolate::SetPromiseRejectCallback(PromiseRejectCallback callback) {
 }
 
 void Isolate::PerformMicrotaskCheckpoint() {
-  DCHECK_NE(MicrotasksPolicy::kScoped, GetMicrotasksPolicy());
+  // DCHECK_NE(MicrotasksPolicy::kScoped, GetMicrotasksPolicy());
   i::Isolate* i_isolate = reinterpret_cast<i::Isolate*>(this);
   i_isolate->default_microtask_queue()->PerformCheckpoint(this);
 }
