From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Shelley Vohr <shelley.vohr@gmail.com>
Date: Mon, 22 Oct 2018 10:47:13 -0700
Subject: expose_mksnapshot.patch

Needed in order to target mksnapshot for mksnapshot zip.

diff --git a/BUILD.gn b/BUILD.gn
index d30d0f6989507066d39a3b5392283e5c48206227..ef3cbdd6426f061ed53fadbc639f0566f799a5e6 100644
--- a/BUILD.gn
+++ b/BUILD.gn
@@ -3844,7 +3844,6 @@ if (current_toolchain == v8_generator_toolchain) {
 
 if (v8_use_snapshot && current_toolchain == v8_snapshot_toolchain) {
   v8_executable("mksnapshot") {
-    visibility = [ ":*" ]  # Only targets in this file can depend on this.
 
     sources = [
       "src/snapshot/embedded/embedded-file-writer.cc",
