From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jeremy Rose <japthorp@slack-corp.com>
Date: Thu, 23 Jun 2022 16:49:28 -0700
Subject: force disable v8 sandboxed pointers

temporarily, until we can fix node to be compatible

diff --git a/BUILD.gn b/BUILD.gn
index 71ee04624868f23a2e204b5ff1e973b45e2f1202..6aa33fdd3c767eb0045db918d2535fe35c98ca4b 100644
--- a/BUILD.gn
+++ b/BUILD.gn
@@ -506,9 +506,9 @@ if (v8_enable_sandbox == "") {
 }
 
 # Enable sandboxed pointers when the sandbox is enabled.
-if (v8_enable_sandbox) {
-  v8_enable_sandboxed_pointers = true
-}
+#if (v8_enable_sandbox) {
+#  v8_enable_sandboxed_pointers = true
+#}
 
 # Enable all available sandbox features if sandbox future is enabled.
 if (v8_enable_sandbox_future) {
