From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jeremy Apthorp <jeremya@chromium.org>
Date: Tue, 17 Sep 2019 18:35:16 +0000
Subject: Revert "[cleanup] Switch {OFFSET_OF} to {offsetof} where possible."

This reverts commit d287e4bc46243841c77cf9798516ee4dcc54bf43.

diff --git a/src/deoptimizer/deoptimizer.h b/src/deoptimizer/deoptimizer.h
index e2cb30fffe5065d1d88d243de439d0251a8797e9..7f1ac45b411b10a658e3287a22b15bc3b8cfba0d 100644
--- a/src/deoptimizer/deoptimizer.h
+++ b/src/deoptimizer/deoptimizer.h
@@ -515,14 +515,14 @@ class Deoptimizer : public Malloced {
                                     DeoptimizeKind* type_out);
 
   // Code generation support.
-  static int input_offset() { return offsetof(Deoptimizer, input_); }
+  static int input_offset() { return OFFSET_OF(Deoptimizer, input_); }
   static int output_count_offset() {
-    return offsetof(Deoptimizer, output_count_);
+    return OFFSET_OF(Deoptimizer, output_count_);
   }
-  static int output_offset() { return offsetof(Deoptimizer, output_); }
+  static int output_offset() { return OFFSET_OF(Deoptimizer, output_); }
 
   static int caller_frame_top_offset() {
-    return offsetof(Deoptimizer, caller_frame_top_);
+    return OFFSET_OF(Deoptimizer, caller_frame_top_);
   }
 
   V8_EXPORT_PRIVATE static int GetDeoptimizedCodeCount(Isolate* isolate);
@@ -791,7 +791,7 @@ class FrameDescription {
   int parameter_count() { return parameter_count_; }
 
   static int registers_offset() {
-    return offsetof(FrameDescription, register_values_.registers_);
+    return OFFSET_OF(FrameDescription, register_values_.registers_);
   }
 
   static constexpr int double_registers_offset() {
