From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Noah Gregory <noahmgregory@gmail.com>
Date: Mon, 2 Mar 2026 02:19:41 -0500
Subject: fix: return early from `CommandBuffer::setEventImpl` if the event is
 `nil`


diff --git a/src/libANGLE/renderer/metal/mtl_command_buffer.mm b/src/libANGLE/renderer/metal/mtl_command_buffer.mm
index 31448a46384b86d411f2249107534527610cec8f..ea4150b6917308c46fa0928ce0e4dcbc12b19789 100644
--- a/src/libANGLE/renderer/metal/mtl_command_buffer.mm
+++ b/src/libANGLE/renderer/metal/mtl_command_buffer.mm
@@ -1195,6 +1195,11 @@ void RandomizeClearValue(ObjCAttachmentDescriptor *objCRenderPassAttachment)
 void CommandBuffer::setEventImpl(id<MTLEvent> event, uint64_t value)
 {
     ASSERT(!getPendingCommandEncoder());
+    // -encodeSignalEvent:value: may crash if the event is nil.
+    if (!event)
+    {
+        return;
+    }
     [get() encodeSignalEvent:event value:value];
 }
 
