From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jeremy Rose <japthorp@slack-corp.com>
Date: Wed, 10 Aug 2022 11:23:38 -0700
Subject: allow base::win::GetDarkModeSupport to block when loading uxtheme.dll

Fixes a DCHECK-on-launch on Windows.

Should be upstreamed.

diff --git a/base/threading/thread_restrictions.h b/base/threading/thread_restrictions.h
index 5455aa07547f753b83dd59c173834034d0673ae7..a59b21b58b5d6a6bbeb9e13214a7021a4c70698a 100644
--- a/base/threading/thread_restrictions.h
+++ b/base/threading/thread_restrictions.h
@@ -382,6 +382,11 @@ class BooleanWithStack;
 
 bool PathProviderWin(int, FilePath*);
 
+namespace win {
+class DarkModeSupport;
+DarkModeSupport& GetDarkModeSupport();
+}
+
 #if DCHECK_IS_ON()
 // NOT_TAIL_CALLED if dcheck-is-on so it's always evident who irrevocably
 // altered the allowance (dcheck-builds will provide the setter's stack on
@@ -498,6 +503,7 @@ class BASE_EXPORT ScopedAllowBlocking {
   friend bool chromeos::system::IsCoreSchedulingAvailable();
   friend int chromeos::system::NumberOfPhysicalCores();
   friend bool disk_cache::CleanupDirectorySync(const base::FilePath&);
+  friend win::DarkModeSupport& win::GetDarkModeSupport();
 
   ScopedAllowBlocking(const Location& from_here = Location::Current());
   ~ScopedAllowBlocking();
diff --git a/base/win/dark_mode_support.cc b/base/win/dark_mode_support.cc
index d15d069b2505e90f2e0c1d85b97440e328e3cbc5..9bf55746f9ddd84ebe86d86cc19add6c2ddc20e4 100644
--- a/base/win/dark_mode_support.cc
+++ b/base/win/dark_mode_support.cc
@@ -8,6 +8,7 @@
 
 #include "base/native_library.h"
 #include "base/no_destructor.h"
+#include "base/threading/thread_restrictions.h"
 #include "base/win/windows_version.h"
 
 namespace {
@@ -61,6 +62,7 @@ const DarkModeSupport& GetDarkModeSupport() {
         // Dark mode only works on WIN10_RS5 and up.
         if (os_info->version() >= base::win::Version::WIN10_RS5) {
           base::NativeLibraryLoadError error;
+          base::ScopedAllowBlocking allow_blocking;
           HMODULE ux_theme_lib = base::PinSystemLibrary(L"uxtheme.dll", &error);
           DCHECK(!error.code);
           if (os_info->version() >= base::win::Version::WIN10_19H1) {
