From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: deepak1556 <hop2deep@gmail.com>
Date: Thu, 21 May 2020 13:58:01 -0700
Subject: Allow setting secondary label via SimpleMenuModel

Builds on https://chromium-review.googlesource.com/c/chromium/src/+/2208976

diff --git a/ui/menus/simple_menu_model.cc b/ui/menus/simple_menu_model.cc
index 7cea2dcb31a7699886a60112beed551c274ff751..1597c84b39826c0dd9be409c312e91409fb69a93 100644
--- a/ui/menus/simple_menu_model.cc
+++ b/ui/menus/simple_menu_model.cc
@@ -57,6 +57,11 @@ std::u16string SimpleMenuModel::Delegate::GetLabelForCommandId(
   return std::u16string();
 }
 
+std::u16string SimpleMenuModel::Delegate::GetSecondaryLabelForCommandId(
+    int command_id) const {
+  return std::u16string();
+}
+
 ImageModel SimpleMenuModel::Delegate::GetIconForCommandId(
     int command_id) const {
   return ImageModel();
@@ -482,6 +487,8 @@ std::u16string SimpleMenuModel::GetLabelAt(size_t index) const {
 }
 
 std::u16string SimpleMenuModel::GetSecondaryLabelAt(size_t index) const {
+  if (IsItemDynamicAt(index))
+    return delegate_->GetSecondaryLabelForCommandId(GetCommandIdAt(index));
   return items_[ValidateItemIndex(index)].secondary_label;
 }
 
diff --git a/ui/menus/simple_menu_model.h b/ui/menus/simple_menu_model.h
index 80f6af93a578edaaa8df335cf6614a0a964ec958..d2a330915454b924893f55612bbbf277c488dca1 100644
--- a/ui/menus/simple_menu_model.h
+++ b/ui/menus/simple_menu_model.h
@@ -100,6 +100,7 @@ class COMPONENT_EXPORT(UI_MENUS) SimpleMenuModel : public MenuModel {
     // Some command ids have labels and icons that change over time.
     virtual bool IsItemForCommandIdDynamic(int command_id) const;
     virtual std::u16string GetLabelForCommandId(int command_id) const;
+    virtual std::u16string GetSecondaryLabelForCommandId(int command_id) const;
     // Gets the icon for the item with the specified id.
     virtual ImageModel GetIconForCommandId(int command_id) const;
 
