From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: deepak1556 <hop2deep@gmail.com>
Date: Thu, 5 Dec 2019 18:56:23 -0800
Subject: base: implement as_const only for libc++

Fixes build error with MSSTL

diff --git a/base/stl_util.h b/base/stl_util.h
index 83d86ad90d24d136bebd46dc02617eea5ec7e438..a8a42c0158cb97c395993fd43caed37f20801e6f 100644
--- a/base/stl_util.h
+++ b/base/stl_util.h
@@ -160,6 +160,7 @@ constexpr const T* data(const std::array<T, N>& array) noexcept {
 
 // C++14 implementation of C++17's std::as_const():
 // https://en.cppreference.com/w/cpp/utility/as_const
+#if defined(_LIBCPP_STD_VER)
 template <typename T>
 constexpr std::add_const_t<T>& as_const(T& t) noexcept {
   return t;
@@ -167,6 +168,9 @@ constexpr std::add_const_t<T>& as_const(T& t) noexcept {
 
 template <typename T>
 void as_const(const T&& t) = delete;
+#else
+using std::as_const;
+#endif
 
 // Returns a const reference to the underlying container of a container adapter.
 // Works for std::priority_queue, std::queue, and std::stack.
