From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: deepak1556 <hop2deep@gmail.com>
Date: Fri, 25 Oct 2019 15:35:45 -0700
Subject: build,[win]: fix ambiguous reference with MSSTL

size if defined in `<xutility>` with VS 2019

diff --git a/base/command_line.cc b/base/command_line.cc
index 4bb290dfa616eb9c7f5f8004eabe33336e9582d5..e9cd0806b1cd58affbfb1d39f7b3fa7a95bec747 100644
--- a/base/command_line.cc
+++ b/base/command_line.cc
@@ -43,7 +43,7 @@ constexpr CommandLine::StringPieceType kSwitchPrefixes[] = {L"--", L"-", L"/"};
 // Unixes don't use slash as a switch.
 constexpr CommandLine::StringPieceType kSwitchPrefixes[] = {"--", "-"};
 #endif
-size_t switch_prefix_count = size(kSwitchPrefixes);
+size_t switch_prefix_count = base::size(kSwitchPrefixes);
 
 size_t GetSwitchPrefixLength(CommandLine::StringPieceType string) {
   for (size_t i = 0; i < switch_prefix_count; ++i) {
@@ -201,7 +201,7 @@ void CommandLine::set_slash_is_not_a_switch() {
   // The last switch prefix should be slash, so adjust the size to skip it.
   static_assert(base::make_span(kSwitchPrefixes).back() == L"/",
                 "Error: Last switch prefix is not a slash.");
-  switch_prefix_count = size(kSwitchPrefixes) - 1;
+  switch_prefix_count = base::size(kSwitchPrefixes) - 1;
 }
 
 // static
