From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samuel Attard <samuel.r.attard@gmail.com>
Date: Wed, 2 Jun 2021 15:10:18 -0700
Subject: build: do not include vr:directx_helpers when !enable_vr

Upstream: https://chromium-review.googlesource.com/c/chromium/src/+/2935426

diff --git a/chrome/test/BUILD.gn b/chrome/test/BUILD.gn
index 4c4d614a47f36d2d75bd83081f10c5d294f59d6c..864708e3f75b3fb8fce7b3cfedda52049c43bcc1 100644
--- a/chrome/test/BUILD.gn
+++ b/chrome/test/BUILD.gn
@@ -4883,9 +4883,7 @@ test("unit_tests") {
     ]
 
     if (!is_electron_build) {
-      deps += [
-        "//chrome:packed_resources_integrity",
-      ]
+      deps += [ "//chrome:packed_resources_integrity" ]
     }
   }
 
@@ -8568,7 +8566,7 @@ if (!is_android) {
       "//third_party/blink/web_tests/resources/testharness.js",
     ]
 
-    if (is_win) {
+    if (is_win && enable_vr) {
       deps += [
         "//device/vr",
         "//device/vr:directx_helpers",
