From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samuel Attard <samuel.r.attard@gmail.com>
Date: Wed, 2 Jun 2021 15:10:18 -0700
Subject: build: do not include vr:directx_helpers when !enable_vr

Upstream: https://chromium-review.googlesource.com/c/chromium/src/+/2935426

diff --git a/chrome/test/BUILD.gn b/chrome/test/BUILD.gn
index 5349ca0848e783db71d02355bd763d2174e4680f..c62fcb965e6c833c0c6a3673e04ce94a4375c673 100644
--- a/chrome/test/BUILD.gn
+++ b/chrome/test/BUILD.gn
@@ -4915,9 +4915,7 @@ test("unit_tests") {
     ]
 
     if (!is_electron_build) {
-      deps += [
-        "//chrome:packed_resources_integrity",
-      ]
+      deps += [ "//chrome:packed_resources_integrity" ]
     }
   }
 
