From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: David Baron <dbaron@chromium.org>
Date: Mon, 24 Mar 2026 11:30:00 -0700
Subject: Check parent nodes when handling vector node insertions.


diff --git a/third_party/blink/renderer/platform/runtime_enabled_features.json5 b/third_party/blink/renderer/platform/runtime_enabled_features.json5
index 80b0abeb7f3c18019aca2f431240a84c6dd749d7..893dec719a744396b17cbed9b9ebbefa540f5ae9 100644
--- a/third_party/blink/renderer/platform/runtime_enabled_features.json5
+++ b/third_party/blink/renderer/platform/runtime_enabled_features.json5
@@ -4375,6 +4375,11 @@
       name: "RecheckParentDuringNodeVectorInsertion",
       status: "stable",
     },
+    {
+      // Shipping in M148, so should be removed around M150.
+      name: "RecheckParentDuringNodeVectorInsertion",
+      status: "stable",
+    },
     {
       name: "RecordSameDocumentPresentationTimeOnce",
       status: "stable",

