From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: deepak1556 <hop2deep@gmail.com>
Date: Wed, 28 Jun 2023 21:11:40 +0900
Subject: chore: add electron to gin embedders

Adds kEmbedderElectron to GinEmbedder for Electron's use.

diff --git a/gin/public/gin_embedders.h b/gin/public/gin_embedders.h
index 6d11f4401d50476802e514392615d287fa54522e..bdd3426b5fa9e94a0eeeb8d81259dc02f46e9bcc 100644
--- a/gin/public/gin_embedders.h
+++ b/gin/public/gin_embedders.h
@@ -20,6 +20,7 @@ enum GinEmbedder : uint16_t {
   kEmbedderBlink,
   kEmbedderPDFium,
   kEmbedderFuchsia,
+  kEmbedderElectron,
 };
 
 enum EmbedderDataTag : uint16_t {
