From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: deepak1556 <hop2deep@gmail.com>
Date: Wed, 28 Jun 2023 21:11:40 +0900
Subject: chore: add electron to gin embedders

Adds kEmbedderElectron to GinEmbedder for Electron's use.

diff --git a/gin/public/gin_embedders.h b/gin/public/gin_embedders.h
index cecf528475cb832ed1876381878eade582bc83d6..2a741def996812148f9007ac9a21336fad1bbd2b 100644
--- a/gin/public/gin_embedders.h
+++ b/gin/public/gin_embedders.h
@@ -20,6 +20,7 @@ enum GinEmbedder : uint16_t {
   kEmbedderBlink,
   kEmbedderPDFium,
   kEmbedderFuchsia,
+  kEmbedderElectron,
 };
 
 enum EmbedderDataTag : uint16_t {
