From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nitish Sakhawalkar <nitsakh@icloud.com>
Date: Mon, 13 May 2019 15:48:36 -0700
Subject: disable_custom_libcxx_on_windows.patch


diff --git a/build/config/c++/c++.gni b/build/config/c++/c++.gni
index 834a943deca28655f40490ab78a14496e253c143..7abf6d83b04b515881c8f797e1031082afe7a5c8 100644
--- a/build/config/c++/c++.gni
+++ b/build/config/c++/c++.gni
@@ -12,7 +12,8 @@ declare_args() {
   # is not supported.
   use_custom_libcxx =
       is_fuchsia || is_android || is_mac || (is_ios && !use_xcode_clang) ||
-      (is_win && is_clang) ||
+      # Do not use custom libcxx on windows
+      #(is_win && is_clang) ||
       (is_linux &&
        (!is_chromeos || default_toolchain != "//build/toolchain/cros:target"))
 
