From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jeremy Apthorp <jeremya@chromium.org>
Date: Fri, 20 Sep 2019 22:56:34 +0000
Subject: fix windows build

Upstream: https://chromium-review.googlesource.com/c/chromium/src/+/1817205

diff --git a/content/browser/frame_host/back_forward_cache_impl.cc b/content/browser/frame_host/back_forward_cache_impl.cc
index cbd6bc1bf77519dd155b911e2a23d599814c37a4..ccf138a0ba2bf18acf6a0b3d4406acf0819cb134 100644
--- a/content/browser/frame_host/back_forward_cache_impl.cc
+++ b/content/browser/frame_host/back_forward_cache_impl.cc
@@ -253,8 +253,9 @@ void BackForwardCacheImpl::PostTaskToDestroyEvictedFrames() {
                                 weak_factory_.GetWeakPtr()));
 }
 
-void BackForwardCacheImpl::DisableForRenderFrameHost(GlobalFrameRoutingId id,
-                                                     std::string_view reason) {
+void BackForwardCacheImpl::DisableForRenderFrameHost(
+    GlobalFrameRoutingId id,
+    const std::string& reason) {
   DCHECK_CURRENTLY_ON(BrowserThread::UI);
   auto* rfh = RenderFrameHostImpl::FromID(id);
   if (rfh)
diff --git a/content/browser/frame_host/back_forward_cache_impl.h b/content/browser/frame_host/back_forward_cache_impl.h
index e81ffc453710ee117bbc7080c3930ad6e1e5ff95..19f4b59c0efac6fc4b9306468b7160f99c4e3375 100644
--- a/content/browser/frame_host/back_forward_cache_impl.h
+++ b/content/browser/frame_host/back_forward_cache_impl.h
@@ -149,7 +149,7 @@ class CONTENT_EXPORT BackForwardCacheImpl : public BackForwardCache {
 
   // BackForwardCache:
   void DisableForRenderFrameHost(GlobalFrameRoutingId id,
-                                 std::string_view reason) override;
+                                 const std::string& reason) override;
 
  private:
   // Destroys all evicted frames in the BackForwardCache.
diff --git a/content/public/browser/back_forward_cache.h b/content/public/browser/back_forward_cache.h
index f7abe8bec74130587b09d66b37f04a7d28ab3705..7e17c4afed0d265fb618dca5cc1bc12b28165ec4 100644
--- a/content/public/browser/back_forward_cache.h
+++ b/content/public/browser/back_forward_cache.h
@@ -5,8 +5,6 @@
 #ifndef CONTENT_PUBLIC_BROWSER_BACK_FORWARD_CACHE_H_
 #define CONTENT_PUBLIC_BROWSER_BACK_FORWARD_CACHE_H_
 
-#include <string_view>
-
 #include "content/common/content_export.h"
 #include "content/public/browser/global_routing_id.h"
 
@@ -43,7 +41,7 @@ class CONTENT_EXPORT BackForwardCache {
   // |id|: If no RenderFrameHost can be found for the given id nothing happens.
   // |reason|: Free form string to be used in logging and metrics.
   virtual void DisableForRenderFrameHost(GlobalFrameRoutingId id,
-                                         std::string_view reason) = 0;
+                                         const std::string& reason) = 0;
 
  protected:
   BackForwardCache() = default;
