From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Sential <sentialx@gmail.com>
Date: Thu, 18 Mar 2021 21:05:59 +0100
Subject: fix: patches for webauthn

Replaces chrome/grit to electron/grit

diff --git a/chrome/browser/ui/views/chrome_typography.cc b/chrome/browser/ui/views/chrome_typography.cc
index de00af3356d6a7bc7f8373d56cab005b262555b3..6b6ef8abde82af93698bfabc2dc577854b8d604d 100644
--- a/chrome/browser/ui/views/chrome_typography.cc
+++ b/chrome/browser/ui/views/chrome_typography.cc
@@ -5,7 +5,6 @@
 #include "chrome/browser/ui/views/chrome_typography.h"
 
 #include "build/build_config.h"
-#include "chrome/browser/ui/views/location_bar/location_bar_view.h"
 #include "ui/base/default_style.h"
 #include "ui/base/pointer/touch_ui_controller.h"
 #include "ui/base/resource/resource_bundle.h"
@@ -75,7 +74,8 @@ void ApplyCommonFontStyles(int context,
     case CONTEXT_OMNIBOX_DEEMPHASIZED: {
       const int omnibox_primary_delta =
           GetFontSizeDeltaBoundedByAvailableHeight(
-              LocationBarView::GetAvailableTextHeight(),
+              // LocationBarView::GetAvailableTextHeight(),
+              ui::TouchUiController::Get()->touch_ui() ? 15 : 14,
               ui::TouchUiController::Get()->touch_ui() ? 15 : 14);
       details.size_delta = omnibox_primary_delta;
       if (context == CONTEXT_OMNIBOX_DEEMPHASIZED)
@@ -90,7 +90,8 @@ void ApplyCommonFontStyles(int context,
       // so don't bother starting at 14.
       const int omnibox_decoration_delta =
           GetFontSizeDeltaBoundedByAvailableHeight(
-              LocationBarView::GetAvailableDecorationTextHeight(), 11);
+              // LocationBarView::GetAvailableDecorationTextHeight()
+              11, 11);
       details.size_delta = omnibox_decoration_delta;
       break;
     }
diff --git a/chrome/browser/ui/views/webauthn/authenticator_client_pin_entry_view.cc b/chrome/browser/ui/views/webauthn/authenticator_client_pin_entry_view.cc
index e689aeb5a828c598d942e183baa9e93913f88b5a..74152bcc45888c067ddaade336c60e87da78d163 100644
--- a/chrome/browser/ui/views/webauthn/authenticator_client_pin_entry_view.cc
+++ b/chrome/browser/ui/views/webauthn/authenticator_client_pin_entry_view.cc
@@ -10,7 +10,7 @@
 #include "base/strings/string_number_conversions.h"
 #include "chrome/browser/ui/views/chrome_layout_provider.h"
 #include "chrome/browser/ui/views/chrome_typography.h"
-#include "chrome/grit/generated_resources.h"
+#include "electron/grit/electron_resources.h"
 #include "ui/base/l10n/l10n_util.h"
 #include "ui/gfx/color_palette.h"
 #include "ui/gfx/geometry/insets.h"
diff --git a/chrome/browser/ui/views/webauthn/authenticator_paask_sheet_view.cc b/chrome/browser/ui/views/webauthn/authenticator_paask_sheet_view.cc
index 0172e32af63370b28d078568b60071fdb12c2e79..0cd85ee3debe5a54051b60332242d21aab51e0f9 100644
--- a/chrome/browser/ui/views/webauthn/authenticator_paask_sheet_view.cc
+++ b/chrome/browser/ui/views/webauthn/authenticator_paask_sheet_view.cc
@@ -7,7 +7,7 @@
 
 #include "chrome/browser/ui/views/webauthn/authenticator_paask_sheet_view.h"
 #include "chrome/browser/ui/webauthn/sheet_models.h"
-#include "chrome/grit/generated_resources.h"
+#include "electron/grit/electron_resources.h"
 #include "ui/base/l10n/l10n_util.h"
 #include "ui/base/resource/resource_bundle.h"
 #include "ui/views/controls/styled_label.h"
diff --git a/chrome/browser/ui/views/webauthn/authenticator_request_dialog_view.cc b/chrome/browser/ui/views/webauthn/authenticator_request_dialog_view.cc
index e4e031c411e316957630b92c7d803771cb1d7fe8..a10c947974075ffc40d6f4d8e30dd8503e95f117 100644
--- a/chrome/browser/ui/views/webauthn/authenticator_request_dialog_view.cc
+++ b/chrome/browser/ui/views/webauthn/authenticator_request_dialog_view.cc
@@ -12,11 +12,11 @@
 #include "chrome/browser/ui/views/webauthn/authenticator_request_sheet_view.h"
 #include "chrome/browser/ui/views/webauthn/sheet_view_factory.h"
 #include "chrome/browser/ui/webauthn/authenticator_request_sheet_model.h"
-#include "chrome/grit/generated_resources.h"
 #include "components/constrained_window/constrained_window_views.h"
 #include "components/strings/grit/components_strings.h"
 #include "components/web_modal/web_contents_modal_dialog_manager.h"
 #include "components/web_modal/web_contents_modal_dialog_manager_delegate.h"
+#include "electron/grit/electron_resources.h"
 #include "ui/base/l10n/l10n_util.h"
 #include "ui/gfx/color_utils.h"
 #include "ui/gfx/geometry/insets.h"
diff --git a/chrome/browser/ui/views/webauthn/authenticator_request_sheet_view.cc b/chrome/browser/ui/views/webauthn/authenticator_request_sheet_view.cc
index 6a7e7ad506efea6afc790ca2511732e1c4bbfab4..aedaaeda9a27b9da68ece657101f5456abe07a3e 100644
--- a/chrome/browser/ui/views/webauthn/authenticator_request_sheet_view.cc
+++ b/chrome/browser/ui/views/webauthn/authenticator_request_sheet_view.cc
@@ -11,9 +11,9 @@
 #include "chrome/browser/ui/views/chrome_typography.h"
 #include "chrome/browser/ui/webauthn/authenticator_request_sheet_model.h"
 #include "chrome/browser/webauthn/authenticator_request_dialog_model.h"
-#include "chrome/grit/generated_resources.h"
 #include "components/strings/grit/components_strings.h"
 #include "components/vector_icons/vector_icons.h"
+#include "electron/grit/electron_resources.h"
 #include "ui/base/l10n/l10n_util.h"
 #include "ui/gfx/color_utils.h"
 #include "ui/gfx/paint_vector_icon.h"
diff --git a/chrome/browser/ui/webauthn/account_hover_list_model.cc b/chrome/browser/ui/webauthn/account_hover_list_model.cc
index 869e26bbcdaafaa0a79ae3822ca1c3302674cb17..ac9f1bdf35d7865f3bedc51081e6039d1540cea3 100644
--- a/chrome/browser/ui/webauthn/account_hover_list_model.cc
+++ b/chrome/browser/ui/webauthn/account_hover_list_model.cc
@@ -6,9 +6,9 @@
 
 #include "base/strings/utf_string_conversions.h"
 #include "chrome/app/vector_icons/vector_icons.h"
-#include "chrome/grit/generated_resources.h"
 #include "device/fido/authenticator_get_assertion_response.h"
 #include "device/fido/public_key_credential_user_entity.h"
+#include "electron/grit/electron_resources.h"
 #include "ui/base/l10n/l10n_util.h"
 #include "ui/gfx/paint_vector_icon.h"
 
diff --git a/chrome/browser/ui/webauthn/other_transports_menu_model.cc b/chrome/browser/ui/webauthn/other_transports_menu_model.cc
index e8b745fe33dcf9b437f23e64aabab93d4d5f4cf1..310530ab66c8d2c166b5502c40b4668518182a00 100644
--- a/chrome/browser/ui/webauthn/other_transports_menu_model.cc
+++ b/chrome/browser/ui/webauthn/other_transports_menu_model.cc
@@ -6,7 +6,7 @@
 
 #include "base/numerics/safe_conversions.h"
 #include "chrome/browser/ui/webauthn/transport_utils.h"
-#include "chrome/grit/generated_resources.h"
+#include "electron/grit/electron_resources.h"
 #include "ui/base/l10n/l10n_util.h"
 #include "ui/base/models/image_model.h"
 #include "ui/gfx/color_palette.h"
diff --git a/chrome/browser/ui/webauthn/sheet_models.cc b/chrome/browser/ui/webauthn/sheet_models.cc
index 052854e3881d5856358278501586cd51119959cd..cb8d84de36834ffe28ac149290dd5188a5701f81 100644
--- a/chrome/browser/ui/webauthn/sheet_models.cc
+++ b/chrome/browser/ui/webauthn/sheet_models.cc
@@ -17,11 +17,11 @@
 #include "build/build_config.h"
 #include "chrome/app/vector_icons/vector_icons.h"
 #include "chrome/browser/ui/webauthn/other_transports_menu_model.h"
-#include "chrome/grit/generated_resources.h"
 #include "components/strings/grit/components_strings.h"
 #include "components/url_formatter/elide_url.h"
 #include "device/fido/authenticator_get_assertion_response.h"
 #include "device/fido/features.h"
+#include "electron/grit/electron_resources.h"
 #include "ui/base/l10n/l10n_util.h"
 #include "ui/gfx/font_list.h"
 #include "ui/gfx/text_utils.h"
diff --git a/chrome/browser/ui/webauthn/transport_hover_list_model.cc b/chrome/browser/ui/webauthn/transport_hover_list_model.cc
index 81a9b9b15049163d2942eb495485d988a20803c2..ab4b81e330971acec5ecd5fe113052b44a8bda6d 100644
--- a/chrome/browser/ui/webauthn/transport_hover_list_model.cc
+++ b/chrome/browser/ui/webauthn/transport_hover_list_model.cc
@@ -8,7 +8,7 @@
 
 #include "chrome/app/vector_icons/vector_icons.h"
 #include "chrome/browser/ui/webauthn/transport_utils.h"
-#include "chrome/grit/generated_resources.h"
+#include "electron/grit/electron_resources.h"
 #include "ui/base/l10n/l10n_util.h"
 #include "ui/gfx/paint_vector_icon.h"
 
diff --git a/chrome/browser/ui/webauthn/transport_utils.cc b/chrome/browser/ui/webauthn/transport_utils.cc
index 312415705b894dcc406414ceae757663ea18a539..4ce48d6654aa5341ea10ebc1947fd17d5ae8eb62 100644
--- a/chrome/browser/ui/webauthn/transport_utils.cc
+++ b/chrome/browser/ui/webauthn/transport_utils.cc
@@ -7,8 +7,8 @@
 #include "base/check_op.h"
 #include "base/notreached.h"
 #include "chrome/app/vector_icons/vector_icons.h"
-#include "chrome/grit/generated_resources.h"
 #include "components/vector_icons/vector_icons.h"
+#include "electron/grit/electron_resources.h"
 #include "ui/base/l10n/l10n_util.h"
 
 namespace {
diff --git a/chrome/browser/webauthn/chrome_authenticator_request_delegate.h b/chrome/browser/webauthn/chrome_authenticator_request_delegate.h
index 49bbabf3d1c3ef01d05164a64dc088d3164fd903..07a88aca44f64d890d48f8545e95449f3b833901 100644
--- a/chrome/browser/webauthn/chrome_authenticator_request_delegate.h
+++ b/chrome/browser/webauthn/chrome_authenticator_request_delegate.h
@@ -22,7 +22,9 @@
 #include "device/fido/fido_request_handler_base.h"
 #include "device/fido/fido_transport_protocol.h"
 
-class Profile;
+namespace electron {
+class ElectronBrowserContext;
+}
 
 namespace content {
 class BrowserContext;
@@ -45,7 +47,7 @@ class ChromeAuthenticatorRequestDelegate
   static void RegisterProfilePrefs(user_prefs::PrefRegistrySyncable* registry);
 #if defined(OS_MAC)
   static TouchIdAuthenticatorConfig TouchIdAuthenticatorConfigForProfile(
-      Profile* profile);
+      electron::ElectronBrowserContext* profile);
 #endif  // defined(OS_MAC)
 
   // The |render_frame_host| must outlive this instance.
