From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Shelley Vohr <shelley.vohr@gmail.com>
Date: Tue, 20 Aug 2019 14:31:53 -0700
Subject: fix: use our base::string16 FromV8


diff --git a/gin/converter.cc b/gin/converter.cc
index 3498080ea5b09eddd7844c313c792f84bcabee1b..8fc421f23a3c720edefcc8402c09ac1d6a5cda5d 100644
--- a/gin/converter.cc
+++ b/gin/converter.cc
@@ -165,13 +165,9 @@ bool Converter<base::string16>::FromV8(Isolate* isolate,
                                        base::string16* out) {
   if (!val->IsString())
     return false;
-  Local<String> str = Local<String>::Cast(val);
-  int length = str->Length();
-  // Note that the reinterpret cast is because on Windows string16 is an alias
-  // to wstring, and hence has character type wchar_t not uint16_t.
-  str->Write(isolate,
-             reinterpret_cast<uint16_t*>(base::WriteInto(out, length + 1)), 0,
-             length);
+  
+  v8::String::Value s(isolate, val);
+  out->assign(reinterpret_cast<const base::char16*>(*s), s.length());
   return true;
 }
 
