From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: George Xu <george.xu@slack-corp.com>
Date: Tue, 25 Jun 2024 13:40:32 -0700
Subject: fixup! fix: workaround frame eviction bug


diff --git a/content/browser/renderer_host/render_widget_host_view_aura.cc b/content/browser/renderer_host/render_widget_host_view_aura.cc
index 75d278e4c440c584ddf18581ae1843c2d0e1f944..220938d3ffccc029ea8d3bd8b8b392087e4c6aac 100644
--- a/content/browser/renderer_host/render_widget_host_view_aura.cc
+++ b/content/browser/renderer_host/render_widget_host_view_aura.cc
@@ -570,6 +570,7 @@ void RenderWidgetHostViewAura::ShowImpl(PageVisibilityState page_visibility) {
   // OnShowWithPageVisibility will not call NotifyHostAndDelegateOnWasShown,
   // which updates `visibility_`, unless the host is hidden. Make sure no update
   // is needed.
+  CHECK(host_->is_hidden() || host->disable_hidden_);
   if (host_->is_hidden() || host()->disable_hidden_ || visibility_ == Visibility::VISIBLE)
     OnShowWithPageVisibility(page_visibility);
 }
