From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jeremy Apthorp <nornagon@nornagon.net>
Date: Thu, 20 Sep 2018 17:49:34 -0700
Subject: mas-lssetapplicationlaunchservicesserverconnectionstatus.patch

Removes usage of the _LSSetApplicationLaunchServicesServerConnectionStatus
private API.

diff --git a/content/gpu/gpu_main.cc b/content/gpu/gpu_main.cc
index 3e3b103d811be454bc9b63541be35949cfd4ec49..49188e7425a2222fac237b998ddff36ed84b61a9 100644
--- a/content/gpu/gpu_main.cc
+++ b/content/gpu/gpu_main.cc
@@ -285,8 +285,10 @@ int GpuMain(const MainFunctionParams& parameters) {
         std::make_unique<base::SingleThreadTaskExecutor>(
             base::MessagePump::Type::NS_RUNLOOP);
 
+#ifndef MAS_BUILD
     // Tell LaunchServices to continue without a connection to the daemon.
     _LSSetApplicationLaunchServicesServerConnectionStatus(0, nullptr);
+#endif
 #else
     main_thread_task_executor =
         std::make_unique<base::SingleThreadTaskExecutor>(
