From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: John Kleinschmidt <jkleinsc@electronjs.org>
Date: Mon, 23 Aug 2021 12:18:19 -0400
Subject: Revert "Roll src/third_party/libgav1/src/ 1c4e344cb..6e13c69d6 (188
 commits)"

This reverts commit 79c0f0423a236b3468d2cb8c1f198114700d2524.

diff --git a/DEPS b/DEPS
index 2864764abeac50de50a66f009105cef5921ad2f5..c9437b93e2009d695085340b20af03f19d2a76b7 100644
--- a/DEPS
+++ b/DEPS
@@ -1078,7 +1078,7 @@ deps = {
     Var('chromium_git') + '/external/github.com/khaledhosny/ots.git' + '@' + Var('ots_revision'),
 
   'src/third_party/libgav1/src':
-    Var('chromium_git') + '/codecs/libgav1.git' + '@' + '6e13c69d6b2507a353e726b5a132ae4a0dd8cca8',
+    Var('chromium_git') + '/codecs/libgav1.git' + '@' + '1c4e344cb1c15aff1c94cb542b98260244a3da4b',
 
   'src/third_party/google_toolbox_for_mac/src': {
       'url': Var('chromium_git') + '/external/github.com/google/google-toolbox-for-mac.git' + '@' + Var('google_toolbox_for_mac_revision'),
diff --git a/third_party/libgav1/README.chromium b/third_party/libgav1/README.chromium
index 7346f4a9c6f0d66427cb93c80e2dea854bf16677..34fdfef4f63467dd75fbc6dce24a67899c97ae75 100644
--- a/third_party/libgav1/README.chromium
+++ b/third_party/libgav1/README.chromium
@@ -2,9 +2,9 @@ Name: libgav1
 Short Name: libgav1
 URL: https://chromium.googlesource.com/codecs/libgav1/
 Version: 0
-Date: Saturday July 24 2021
+Date: Friday March 12 2021
 Branch: master
-Commit: 6e13c69d6b2507a353e726b5a132ae4a0dd8cca8
+Commit: 1c4e344cb1c15aff1c94cb542b98260244a3da4b
 License: Apache 2.0
 License File: libgav1/LICENSE
 Security Critical: yes
diff --git a/third_party/libgav1/libgav1_srcs.gni b/third_party/libgav1/libgav1_srcs.gni
index 8061613c99077eb6cb469552eba80fd093043203..abcb887e781762f3155ca6c344453c5da7e70267 100644
--- a/third_party/libgav1/libgav1_srcs.gni
+++ b/third_party/libgav1/libgav1_srcs.gni
@@ -44,20 +44,18 @@ gav1_common_sources = [
 gav1_dsp_sources = [
   "//third_party/libgav1/src/src/dsp/arm/average_blend_neon.cc",
   "//third_party/libgav1/src/src/dsp/arm/cdef_neon.cc",
-  "//third_party/libgav1/src/src/dsp/arm/convolve_10bit_neon.cc",
   "//third_party/libgav1/src/src/dsp/arm/convolve_neon.cc",
   "//third_party/libgav1/src/src/dsp/arm/distance_weighted_blend_neon.cc",
   "//third_party/libgav1/src/src/dsp/arm/film_grain_neon.cc",
   "//third_party/libgav1/src/src/dsp/arm/intra_edge_neon.cc",
   "//third_party/libgav1/src/src/dsp/arm/intrapred_cfl_neon.cc",
   "//third_party/libgav1/src/src/dsp/arm/intrapred_directional_neon.cc",
-  "//third_party/libgav1/src/src/dsp/arm/intrapred_filter_neon.cc",
+  "//third_party/libgav1/src/src/dsp/arm/intrapred_filter_intra_neon.cc",
   "//third_party/libgav1/src/src/dsp/arm/intrapred_neon.cc",
   "//third_party/libgav1/src/src/dsp/arm/intrapred_smooth_neon.cc",
   "//third_party/libgav1/src/src/dsp/arm/inverse_transform_10bit_neon.cc",
   "//third_party/libgav1/src/src/dsp/arm/inverse_transform_neon.cc",
   "//third_party/libgav1/src/src/dsp/arm/loop_filter_neon.cc",
-  "//third_party/libgav1/src/src/dsp/arm/loop_restoration_10bit_neon.cc",
   "//third_party/libgav1/src/src/dsp/arm/loop_restoration_neon.cc",
   "//third_party/libgav1/src/src/dsp/arm/mask_blend_neon.cc",
   "//third_party/libgav1/src/src/dsp/arm/motion_field_projection_neon.cc",
@@ -75,10 +73,6 @@ gav1_dsp_sources = [
   "//third_party/libgav1/src/src/dsp/film_grain.cc",
   "//third_party/libgav1/src/src/dsp/intra_edge.cc",
   "//third_party/libgav1/src/src/dsp/intrapred.cc",
-  "//third_party/libgav1/src/src/dsp/intrapred_cfl.cc",
-  "//third_party/libgav1/src/src/dsp/intrapred_directional.cc",
-  "//third_party/libgav1/src/src/dsp/intrapred_filter.cc",
-  "//third_party/libgav1/src/src/dsp/intrapred_smooth.cc",
   "//third_party/libgav1/src/src/dsp/inverse_transform.cc",
   "//third_party/libgav1/src/src/dsp/loop_filter.cc",
   "//third_party/libgav1/src/src/dsp/loop_restoration.cc",
@@ -113,11 +107,7 @@ gav1_dsp_headers_sources = [
   "//third_party/libgav1/src/src/dsp/arm/distance_weighted_blend_neon.h",
   "//third_party/libgav1/src/src/dsp/arm/film_grain_neon.h",
   "//third_party/libgav1/src/src/dsp/arm/intra_edge_neon.h",
-  "//third_party/libgav1/src/src/dsp/arm/intrapred_cfl_neon.h",
-  "//third_party/libgav1/src/src/dsp/arm/intrapred_directional_neon.h",
-  "//third_party/libgav1/src/src/dsp/arm/intrapred_filter_neon.h",
   "//third_party/libgav1/src/src/dsp/arm/intrapred_neon.h",
-  "//third_party/libgav1/src/src/dsp/arm/intrapred_smooth_neon.h",
   "//third_party/libgav1/src/src/dsp/arm/inverse_transform_neon.h",
   "//third_party/libgav1/src/src/dsp/arm/loop_filter_neon.h",
   "//third_party/libgav1/src/src/dsp/arm/loop_restoration_neon.h",
@@ -139,10 +129,6 @@ gav1_dsp_headers_sources = [
   "//third_party/libgav1/src/src/dsp/film_grain_common.h",
   "//third_party/libgav1/src/src/dsp/intra_edge.h",
   "//third_party/libgav1/src/src/dsp/intrapred.h",
-  "//third_party/libgav1/src/src/dsp/intrapred_cfl.h",
-  "//third_party/libgav1/src/src/dsp/intrapred_directional.h",
-  "//third_party/libgav1/src/src/dsp/intrapred_filter.h",
-  "//third_party/libgav1/src/src/dsp/intrapred_smooth.h",
   "//third_party/libgav1/src/src/dsp/inverse_transform.h",
   "//third_party/libgav1/src/src/dsp/loop_filter.h",
   "//third_party/libgav1/src/src/dsp/loop_restoration.h",
@@ -163,8 +149,6 @@ gav1_dsp_sse4_sources = [
   "//third_party/libgav1/src/src/dsp/x86/film_grain_sse4.cc",
   "//third_party/libgav1/src/src/dsp/x86/intra_edge_sse4.cc",
   "//third_party/libgav1/src/src/dsp/x86/intrapred_cfl_sse4.cc",
-  "//third_party/libgav1/src/src/dsp/x86/intrapred_directional_sse4.cc",
-  "//third_party/libgav1/src/src/dsp/x86/intrapred_filter_sse4.cc",
   "//third_party/libgav1/src/src/dsp/x86/intrapred_smooth_sse4.cc",
   "//third_party/libgav1/src/src/dsp/x86/intrapred_sse4.cc",
   "//third_party/libgav1/src/src/dsp/x86/inverse_transform_sse4.cc",
@@ -188,10 +172,6 @@ gav1_dsp_sse4_headers_sources = [
   "//third_party/libgav1/src/src/dsp/x86/distance_weighted_blend_sse4.h",
   "//third_party/libgav1/src/src/dsp/x86/film_grain_sse4.h",
   "//third_party/libgav1/src/src/dsp/x86/intra_edge_sse4.h",
-  "//third_party/libgav1/src/src/dsp/x86/intrapred_cfl_sse4.h",
-  "//third_party/libgav1/src/src/dsp/x86/intrapred_directional_sse4.h",
-  "//third_party/libgav1/src/src/dsp/x86/intrapred_filter_sse4.h",
-  "//third_party/libgav1/src/src/dsp/x86/intrapred_smooth_sse4.h",
   "//third_party/libgav1/src/src/dsp/x86/intrapred_sse4.h",
   "//third_party/libgav1/src/src/dsp/x86/inverse_transform_sse4.h",
   "//third_party/libgav1/src/src/dsp/x86/loop_filter_sse4.h",
