From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: John Kleinschmidt <jkleinsc@electronjs.org>
Date: Mon, 23 Aug 2021 13:42:34 -0400
Subject: Revert "Roll clang llvmorg-14-init-1002-gb5e470aa-1 :
 llvmorg-14-init-1380-gee659383-1"

This reverts commit 65172c258c45f5bc22f2018884d013b8d11f4b26.

diff --git a/tools/clang/scripts/update.py b/tools/clang/scripts/update.py
index 56b1a39e7883c89fbf27a205c136fc2c12a785f5..64f27320cd0f99934ff0c2e105a7cb08a53b5c88 100755
--- a/tools/clang/scripts/update.py
+++ b/tools/clang/scripts/update.py
@@ -39,7 +39,7 @@ import zipfile
 # https://chromium.googlesource.com/chromium/src/+/main/docs/updating_clang.md
 # Reverting problematic clang rolls is safe, though.
 # This is the output of `git describe` and is usable as a commit-ish.
-CLANG_REVISION = 'llvmorg-14-init-1380-gee659383'
+CLANG_REVISION = 'llvmorg-14-init-1002-gb5e470aa'
 CLANG_SUB_REVISION = 1
 
 PACKAGE_VERSION = '%s-%s' % (CLANG_REVISION, CLANG_SUB_REVISION)
