From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Cheng Zhao <zcbenz@gmail.com>
Date: Thu, 20 Sep 2018 17:49:03 -0700
Subject: scroll_bounce_flag.patch

Patch to make scrollBounce option work.

diff --git a/content/renderer/render_thread_impl.cc b/content/renderer/render_thread_impl.cc
index b15d81e09742c135350d0a32e3e68c810478c3b7..66304031e2b2dd6b4f9b1fc8f21229da80446ceb 100644
--- a/content/renderer/render_thread_impl.cc
+++ b/content/renderer/render_thread_impl.cc
@@ -1354,7 +1354,7 @@ bool RenderThreadImpl::IsGpuMemoryBufferCompositorResourcesEnabled() {
 }
 
 bool RenderThreadImpl::IsElasticOverscrollEnabled() {
-  return is_elastic_overscroll_enabled_;
+  return base::CommandLine::ForCurrentProcess()->HasSwitch("scroll-bounce");
 }
 
 bool RenderThreadImpl::IsUseZoomForDSFEnabled() {
