From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Cheng Zhao <zcbenz@gmail.com>
Date: Thu, 20 Sep 2018 17:49:03 -0700
Subject: scroll_bounce_flag.patch

Patch to make scrollBounce option work.

diff --git a/content/renderer/render_thread_impl.cc b/content/renderer/render_thread_impl.cc
index 95ab5d2de52ec021867151c8de30c78cd4828a20..a20cbbc2d2f535d64b63c15f04c6f6aaddcec413 100644
--- a/content/renderer/render_thread_impl.cc
+++ b/content/renderer/render_thread_impl.cc
@@ -1251,7 +1251,7 @@ bool RenderThreadImpl::IsLcdTextEnabled() {
 }
 
 bool RenderThreadImpl::IsElasticOverscrollEnabled() {
-  return is_elastic_overscroll_enabled_;
+  return base::CommandLine::ForCurrentProcess()->HasSwitch("scroll-bounce");
 }
 
 gpu::GpuMemoryBufferManager* RenderThreadImpl::GetGpuMemoryBufferManager() {
