From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Cheng Zhao <zcbenz@gmail.com>
Date: Thu, 20 Sep 2018 17:49:03 -0700
Subject: scroll_bounce_flag.patch

Patch to make scrollBounce option work.

diff --git a/content/renderer/render_thread_impl.cc b/content/renderer/render_thread_impl.cc
index fcb90d8024c7677ea15ff73064f51886b333a512..fa2ecae71873d4d83b805f73b8978fbadb5b1245 100644
--- a/content/renderer/render_thread_impl.cc
+++ b/content/renderer/render_thread_impl.cc
@@ -1270,7 +1270,7 @@ bool RenderThreadImpl::IsLcdTextEnabled() {
 }
 
 bool RenderThreadImpl::IsElasticOverscrollEnabled() {
-  return is_elastic_overscroll_enabled_;
+  return base::CommandLine::ForCurrentProcess()->HasSwitch("scroll-bounce");
 }
 
 bool RenderThreadImpl::IsUseZoomForDSFEnabled() {
