From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: deepak1556 <hop2deep@gmail.com>
Date: Fri, 10 Apr 2020 17:47:18 -0700
Subject: ui_gtk_public_header.patch

Allow electron to depend on //ui/gtk/gtk_util.h

diff --git a/ui/gtk/BUILD.gn b/ui/gtk/BUILD.gn
index 6ff955727e13551f59ca06f7acf2a4d3a5ab0dc3..98113800c75d2a0dfc86b5ee21040dfef3fd7c05 100644
--- a/ui/gtk/BUILD.gn
+++ b/ui/gtk/BUILD.gn
@@ -69,7 +69,7 @@ generate_stubs("gtk_stubs") {
 }
 
 component("gtk") {
-  public = [ "gtk_ui_factory.h" ]
+  public = [ "gtk_ui_factory.h", "gtk_util.h" ]
 
   sources = [
     "gtk_color_mixers.cc",
@@ -83,7 +83,6 @@ component("gtk") {
     "gtk_ui_factory.cc",
     "gtk_ui_platform.h",
     "gtk_util.cc",
-    "gtk_util.h",
     "input_method_context_impl_gtk.cc",
     "input_method_context_impl_gtk.h",
     "native_theme_gtk.cc",
