From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: deepak1556 <hop2deep@gmail.com>
Date: Fri, 10 Apr 2020 17:47:18 -0700
Subject: ui_gtk_public_header.patch

Allow electron to depend on //ui/gtk/gtk_util.h

diff --git a/ui/gtk/BUILD.gn b/ui/gtk/BUILD.gn
index 73d65668a65802b0ca6112d05c76aefddd157502..b4f29ffa2e029cbcee413aaa7bf54cc03e790864 100644
--- a/ui/gtk/BUILD.gn
+++ b/ui/gtk/BUILD.gn
@@ -39,11 +39,13 @@ component("gtk_ui_delegate") {
 }
 
 component("gtk") {
-  public = [ "gtk_ui.h" ]
+  public = [
+    "gtk_ui.h",
+    "gtk_util.h",
+  ]
   sources = [
     "gtk_ui.cc",
     "gtk_util.cc",
-    "gtk_util.h",
     "input_method_context_impl_gtk.cc",
     "input_method_context_impl_gtk.h",
     "native_theme_gtk.cc",
