From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: deepak1556 <hop2deep@gmail.com>
Date: Fri, 10 Apr 2020 17:47:18 -0700
Subject: ui_gtk_public_header.patch

Allow electron to depend on //ui/gtk/gtk_util.h

diff --git a/ui/gtk/BUILD.gn b/ui/gtk/BUILD.gn
index 6cf6871d3748177d247ee10d5689b97b7dcf0a17..d6a5598004c7ae46542e47953551847a6995c8d3 100644
--- a/ui/gtk/BUILD.gn
+++ b/ui/gtk/BUILD.gn
@@ -68,7 +68,7 @@ generate_stubs("gtk_stubs") {
 }
 
 component("gtk") {
-  public = [ "gtk_ui_factory.h" ]
+  public = [ "gtk_ui_factory.h", "gtk_util.h" ]
 
   sources = [
     "gtk_color_mixers.cc",
@@ -82,7 +82,6 @@ component("gtk") {
     "gtk_ui_factory.cc",
     "gtk_ui_platform.h",
     "gtk_util.cc",
-    "gtk_util.h",
     "input_method_context_impl_gtk.cc",
     "input_method_context_impl_gtk.h",
     "native_theme_gtk.cc",
