From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: deepak1556 <hop2deep@gmail.com>
Date: Fri, 10 Apr 2020 17:47:18 -0700
Subject: ui_gtk_public_header.patch

Allow electron to depend on //ui/gtk/gtk_util.h

diff --git a/ui/gtk/BUILD.gn b/ui/gtk/BUILD.gn
index 4093df78da0bbb1d8df743942f364cf728ad3414..2f31d99b207ffc3531b5334b5a01239cc1fefb35 100644
--- a/ui/gtk/BUILD.gn
+++ b/ui/gtk/BUILD.gn
@@ -69,7 +69,7 @@ generate_stubs("gtk_stubs") {
 }
 
 component("gtk") {
-  public = [ "gtk_ui_factory.h" ]
+  public = [ "gtk_ui_factory.h", "gtk_util.h" ]
 
   sources = [
     "gtk_color_mixers.cc",
@@ -85,7 +85,6 @@ component("gtk") {
     "gtk_ui_platform_stub.cc",
     "gtk_ui_platform_stub.h",
     "gtk_util.cc",
-    "gtk_util.h",
     "input_method_context_impl_gtk.cc",
     "input_method_context_impl_gtk.h",
     "native_theme_gtk.cc",
