From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: deepak1556 <hop2deep@gmail.com>
Date: Fri, 10 Apr 2020 17:47:18 -0700
Subject: ui_gtk_public_header.patch

Allow electron to depend on //ui/gtk/gtk_util.h

diff --git a/ui/gtk/BUILD.gn b/ui/gtk/BUILD.gn
index b982ebc25ed25fe3013e4b1dd39607b7011dd0b6..ad629e388224c2e420a6329723540eed2d6fd2ab 100644
--- a/ui/gtk/BUILD.gn
+++ b/ui/gtk/BUILD.gn
@@ -55,7 +55,10 @@ component("gtk_ui_delegate") {
 }
 
 component("gtk") {
-  public = [ "gtk_ui.h" ]
+  public = [
+    "gtk_ui.h",
+    "gtk_util.h",
+  ]
 
   friend = [ ":gtk_unittests" ]
 
@@ -66,7 +69,6 @@ component("gtk") {
     "gtk_key_bindings_handler.h",
     "gtk_ui.cc",
     "gtk_util.cc",
-    "gtk_util.h",
     "input_method_context_impl_gtk.cc",
     "input_method_context_impl_gtk.h",
     "native_theme_gtk.cc",
