From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: deepak1556 <hop2deep@gmail.com>
Date: Fri, 10 Apr 2020 17:47:18 -0700
Subject: ui_gtk_public_header.patch

Allow electron to depend on //ui/gtk/gtk_util.h

diff --git a/ui/gtk/BUILD.gn b/ui/gtk/BUILD.gn
index 0281dd4b859af32f70db378800318bbfd0b798da..aac8eef61ef7e4c35eda1b7e2fdd8235063195ae 100644
--- a/ui/gtk/BUILD.gn
+++ b/ui/gtk/BUILD.gn
@@ -66,7 +66,10 @@ component("gtk_ui_delegate") {
 }
 
 component("gtk") {
-  public = [ "gtk_ui.h" ]
+  public = [
+    "gtk_ui.h",
+    "gtk_util.h",
+  ]
 
   friend = [ ":gtk_unittests" ]
 
@@ -77,7 +80,6 @@ component("gtk") {
     "gtk_key_bindings_handler.h",
     "gtk_ui.cc",
     "gtk_util.cc",
-    "gtk_util.h",
     "input_method_context_impl_gtk.cc",
     "input_method_context_impl_gtk.h",
     "native_theme_gtk.cc",
