From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: deepak1556 <hop2deep@gmail.com>
Date: Fri, 10 Apr 2020 17:47:18 -0700
Subject: ui_gtk_public_header.patch

Allow electron to depend on //ui/gtk/gtk_util.h

diff --git a/ui/gtk/BUILD.gn b/ui/gtk/BUILD.gn
index 99d97e9c14532d02d2127d3e401105da758bb1a5..1613f89030549260b5b9879eea067ee83223bade 100644
--- a/ui/gtk/BUILD.gn
+++ b/ui/gtk/BUILD.gn
@@ -22,13 +22,15 @@ component("gtk_ui_delegate") {
 }
 
 jumbo_component("gtk") {
-  public = [ "gtk_ui.h" ]
+  public = [ 
+    "gtk_ui.h",
+    "gtk_util.h",
+  ]
   sources = [
     "gtk_key_bindings_handler.cc",
     "gtk_key_bindings_handler.h",
     "gtk_ui.cc",
     "gtk_util.cc",
-    "gtk_util.h",
     "input_method_context_impl_gtk.cc",
     "input_method_context_impl_gtk.h",
     "native_theme_gtk.cc",
