From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jeremy Apthorp <nornagon@nornagon.net>
Date: Tue, 16 Apr 2019 11:25:08 -0700
Subject: unsandboxed_ppapi_processes_skip_zygote.patch

Unsandboxed ppapi processes should skip zygote.

diff --git a/content/browser/ppapi_plugin_process_host.cc b/content/browser/ppapi_plugin_process_host.cc
index ae0a8f39952a4cfb90393d3bb00bed494ce8c5f1..dd85840f1294e2f9f2063a2d24fe372bd571c723 100644
--- a/content/browser/ppapi_plugin_process_host.cc
+++ b/content/browser/ppapi_plugin_process_host.cc
@@ -111,6 +111,9 @@ class PpapiPluginSandboxedProcessLauncherDelegate
   ZygoteHandle GetZygote() override {
     const base::CommandLine& browser_command_line =
         *base::CommandLine::ForCurrentProcess();
+    if (browser_command_line.HasSwitch(sandbox::policy::switches::kNoSandbox)) {
+      return nullptr;
+    }
     base::CommandLine::StringType plugin_launcher = browser_command_line
         .GetSwitchValueNative(switches::kPpapiPluginLauncher);
     if (!plugin_launcher.empty())
