From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tomasz <tomasz@openfin.co>
Date: Fri, 17 Sep 2021 14:18:11 -0700
Subject: Fix proablem with "GetHostNameW" on Windows 7.


diff --git a/deps/uv/src/win/util.c b/deps/uv/src/win/util.c
index 88602c7ee8..aad8f1a15e 100644
--- a/deps/uv/src/win/util.c
+++ b/deps/uv/src/win/util.c
@@ -1664,33 +1664,26 @@ int uv_os_unsetenv(const char* name) {


 int uv_os_gethostname(char* buffer, size_t* size) {
-  WCHAR buf[UV_MAXHOSTNAMESIZE];
+  char buf[UV_MAXHOSTNAMESIZE];
   size_t len;
-  char* utf8_str;
-  int convert_result;

   if (buffer == NULL || size == NULL || *size == 0)
     return UV_EINVAL;

   uv__once_init(); /* Initialize winsock */

-  if (GetHostNameW(buf, UV_MAXHOSTNAMESIZE) != 0)
+  if (gethostname(buf, sizeof(buf)) != 0)
     return uv_translate_sys_error(WSAGetLastError());

-  convert_result = uv__convert_utf16_to_utf8(buf, -1, &utf8_str);
+  buf[sizeof(buf) - 1] = '\0'; /* Null terminate, just to be safe. */
+  len = strlen(buf);

-  if (convert_result != 0)
-    return convert_result;
-
-  len = strlen(utf8_str);
   if (len >= *size) {
     *size = len + 1;
-    uv__free(utf8_str);
     return UV_ENOBUFS;
   }

-  memcpy(buffer, utf8_str, len + 1);
-  uv__free(utf8_str);
+  memcpy(buffer, buf, len + 1);
   *size = len;
   return 0;
 }
