From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jeremy Apthorp <nornagon@nornagon.net>
Date: Tue, 2 Oct 2018 17:09:15 -0700
Subject: export Environment::kNodeContextTagPtr


diff --git a/src/env.h b/src/env.h
index b6aa8e9996ed5c76e9e0c1324d3a9d063246d456..8eed574f52410ddb772f4e075b9cf9a8259a7a4c 100644
--- a/src/env.h
+++ b/src/env.h
@@ -1301,7 +1301,7 @@ class Environment : public MemoryRetainer {
   uint64_t thread_id_;
   std::unordered_set<worker::Worker*> sub_worker_contexts_;
 
-  static void* const kNodeContextTagPtr;
+  NODE_EXTERN static void* const kNodeContextTagPtr;
   static int const kNodeContextTag;
 
 #if HAVE_INSPECTOR
