From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Shelley Vohr <shelley.vohr@gmail.com>
Date: Thu, 13 Sep 2018 08:56:07 -0700
Subject: feat: initialize asar support

This patch initializes asar support in Node.js.

diff --git a/lib/internal/process/pre_execution.js b/lib/internal/process/pre_execution.js
index 23d4dbcf6cd8e1e3e4a509672e203323d81e736c..a951a526a1901ec8d83f63df5de8c8f3ed732667 100644
--- a/lib/internal/process/pre_execution.js
+++ b/lib/internal/process/pre_execution.js
@@ -103,12 +103,17 @@ function prepareMainThreadExecution(expandArgv1 = false,
   assert(!CJSLoader.hasLoadedAnyUserCJSModule);
   loadPreloadModules();
   initializeFrozenIntrinsics();
+  setupAsarSupport();
 }
 
 function refreshRuntimeOptions() {
   refreshOptions();
 }
 
+function setupAsarSupport() {
+  process._linkedBinding('electron_common_asar').initAsarSupport(require);
+}
+
 function patchProcessObject(expandArgv1) {
   const binding = internalBinding('process_methods');
   binding.patchProcessObject(process);

