From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Shelley Vohr <shelley.vohr@gmail.com>
Date: Thu, 13 Sep 2018 08:56:07 -0700
Subject: feat: initialize asar support

This patch initializes asar support in Node.js.

diff --git a/lib/internal/process/pre_execution.js b/lib/internal/process/pre_execution.js
index 250a43c5455b4f9ff72dd3a34d5b0aa270f43cc6..2e35cbba2ac02494c44821af395fe0195b1ab6b5 100644
--- a/lib/internal/process/pre_execution.js
+++ b/lib/internal/process/pre_execution.js
@@ -117,12 +117,17 @@ function setupUserModules() {
   loadPreloadModules();
   // Need to be done after --require setup.
   initializeFrozenIntrinsics();
+  setupAsarSupport();
 }
 
 function refreshRuntimeOptions() {
   refreshOptions();
 }
 
+function setupAsarSupport() {
+  process._linkedBinding('electron_common_asar').initAsarSupport(require);
+}
+
 function patchProcessObject(expandArgv1) {
   const binding = internalBinding('process_methods');
   binding.patchProcessObject(process);
