From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nitish Sakhawalkar <nitsakh@icloud.com>
Date: Wed, 27 Mar 2019 14:17:15 -0700
Subject: fix: Export node::Abort and Assert


diff --git a/src/node_errors.cc b/src/node_errors.cc
index 603c9e415c3f546c621793afd13ee58bcc4280be..20098777e290bfa66dc430c08eaad78e4c232266 100644
--- a/src/node_errors.cc
+++ b/src/node_errors.cc
@@ -241,13 +241,13 @@ void AppendExceptionLine(Environment* env,
             .FromMaybe(false));
 }
 
-[[noreturn]] void Abort() {
+[[noreturn]] void NODE_EXTERN Abort() {
   DumpBacktrace(stderr);
   fflush(stderr);
   ABORT_NO_BACKTRACE();
 }
 
-[[noreturn]] void Assert(const AssertionInfo& info) {
+[[noreturn]] void NODE_EXTERN Assert(const AssertionInfo& info) {
   char name[1024];
   GetHumanReadableProcessName(&name);
 
diff --git a/src/util.h b/src/util.h
index a94e88f232fd3116fa842aed4d02a78fd0c54ffc..62ae0fe6e6beb77e456bff352246505531ec2005 100644
--- a/src/util.h
+++ b/src/util.h
@@ -24,6 +24,8 @@
 
 #if defined(NODE_WANT_INTERNALS) && NODE_WANT_INTERNALS
 
+
+#include "node.h"
 #include "v8.h"
 
 #include <cassert>
@@ -103,8 +105,8 @@ struct AssertionInfo {
   const char* message;
   const char* function;
 };
-[[noreturn]] void Assert(const AssertionInfo& info);
-[[noreturn]] void Abort();
+[[noreturn]] void NODE_EXTERN Assert(const AssertionInfo& info);
+[[noreturn]] void NODE_EXTERN Abort();
 void DumpBacktrace(FILE* fp);
 
 // Windows 8+ does not like abort() in Release mode
