From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: deepak1556 <hop2deep@gmail.com>
Date: Fri, 20 Aug 2021 22:41:11 -0700
Subject: fix: -Wunreachable-code-return

Should be upstreamed.

diff --git a/src/node_url.cc b/src/node_url.cc
index d7549e3bc05562d67bdef5aebf1fefa550b72eb6..d78cb7a3e1ceb075bd532e823a915b18f5a52afd 100644
--- a/src/node_url.cc
+++ b/src/node_url.cc
@@ -600,7 +600,6 @@ std::string URLHost::ToString() const {
     case HostType::H_DOMAIN:
     case HostType::H_OPAQUE:
       return value_.domain_or_opaque;
-      break;
     case HostType::H_IPV4: {
       dest.reserve(15);
       uint32_t value = value_.ipv4;
