From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jeremy Rose <nornagon@nornagon.net>
Date: Fri, 28 Jan 2022 13:50:21 -0800
Subject: fix serdes test

The V8 wire format version changed.

diff --git a/test/parallel/test-v8-serdes.js b/test/parallel/test-v8-serdes.js
index 1b6638ac1a90bdcec618b6c7b2a51c21fe6b548e..5ce5668925f0defb685d77063212f42eb524b3ad 100644
--- a/test/parallel/test-v8-serdes.js
+++ b/test/parallel/test-v8-serdes.js
@@ -164,11 +164,11 @@ const hostObject = new (internalBinding('js_stream').JSStream)();
 
 {
   // Test that an old serialized value can still be deserialized.
-  const buf = Buffer.from('ff0d6f2203666f6f5e007b01', 'hex');
+  const buf = Buffer.from('ff0f6f2203666f6f5e007b01', 'hex');
 
   const des = new v8.DefaultDeserializer(buf);
   des.readHeader();
-  assert.strictEqual(des.getWireFormatVersion(), 0x0d);
+  assert.strictEqual(des.getWireFormatVersion(), 0x0f);
 
   const value = des.readValue();
   assert.strictEqual(value, value.foo);
@@ -203,7 +203,7 @@ const hostObject = new (internalBinding('js_stream').JSStream)();
 {
   // Unaligned Uint16Array read, with padding in the underlying array buffer.
   let buf = Buffer.alloc(32 + 9);
-  buf.write('ff0d5c0404addeefbe', 32, 'hex');
+  buf.write('ff0e5c0404addeefbe', 32, 'hex');
   buf = buf.slice(32);
 
   const expectedResult = os.endianness() === 'LE' ?
