From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jeremy Rose <nornagon@nornagon.net>
Date: Fri, 28 Jan 2022 13:50:21 -0800
Subject: fix serdes test

The V8 wire format version changed.

diff --git a/test/parallel/test-v8-serdes.js b/test/parallel/test-v8-serdes.js
index ef9ef5945dba3b0748d5d0671f87eb20984de3c4..12f20ed1c9d386122dd20fdd84a7a0c9b9079ee1 100644
--- a/test/parallel/test-v8-serdes.js
+++ b/test/parallel/test-v8-serdes.js
@@ -163,7 +163,7 @@ const hostObject = new (internalBinding('js_stream').JSStream)();
 }
 
 {
-  const buf = Buffer.from('ff0d6f2203666f6f5e007b01', 'hex');
+  const buf = Buffer.from('ff0e6f2203666f6f5e007b01', 'hex');
 
   const des = new v8.DefaultDeserializer(buf);
   des.readHeader();
@@ -174,13 +174,13 @@ const hostObject = new (internalBinding('js_stream').JSStream)();
   ser.writeValue(des.readValue());
 
   assert.deepStrictEqual(buf, ser.releaseBuffer());
-  assert.strictEqual(des.getWireFormatVersion(), 0x0d);
+  assert.strictEqual(des.getWireFormatVersion(), 0x0e);
 }
 
 {
   // Unaligned Uint16Array read, with padding in the underlying array buffer.
   let buf = Buffer.alloc(32 + 9);
-  buf.write('ff0d5c0404addeefbe', 32, 'hex');
+  buf.write('ff0e5c0404addeefbe', 32, 'hex');
   buf = buf.slice(32);
 
   const expectedResult = os.endianness() === 'LE' ?
