From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: deepak1556 <hop2deep@gmail.com>
Date: Tue, 16 Nov 2021 20:05:57 +0900
Subject: fix: suppress clang -Wdeprecated-declarations in libuv

Should be upstreamed.

diff --git a/deps/uv/src/win/util.c b/deps/uv/src/win/util.c
index 056dd1e28ea68f5c363fa36dac175743d6cd0ccb..c87719a57fd925b1fb58335c5632322b5533329d 100644
--- a/deps/uv/src/win/util.c
+++ b/deps/uv/src/win/util.c
@@ -1956,10 +1956,17 @@ int uv_os_uname(uv_utsname_t* buffer) {
     #ifdef _MSC_VER
     #pragma warning(suppress : 4996)
     #endif
+    #ifdef __clang__
+    #pragma clang diagnostic push
+    #pragma clang diagnostic ignored "-Wdeprecated-declarations"
+    #endif
     if (GetVersionExW(&os_info) == 0) {
       r = uv_translate_sys_error(GetLastError());
       goto error;
     }
+    #ifdef __clang__
+    #pragma clang diagnostic pop
+    #endif
   }
 
   /* Populate the version field. */
