From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Shelley Vohr <shelley.vohr@gmail.com>
Date: Tue, 22 Oct 2019 11:00:01 -0700
Subject: fix: call InitializeContextRuntime in InitializeContext


diff --git a/src/api/environment.cc b/src/api/environment.cc
index ae26cb7e9ef55fc0b965e28de4686aec87f42522..05282ebb4414c9f1757763fd68893bf0c6005279 100644
--- a/src/api/environment.cc
+++ b/src/api/environment.cc
@@ -362,8 +362,6 @@ Local<Context> NewContext(Isolate* isolate,
     return Local<Context>();
   }
 
-  InitializeContextRuntime(context);
-
   return context;
 }
 
@@ -448,6 +446,11 @@ bool InitializeContext(Local<Context> context) {
     }
   }
 
+  // This must be called here because embedders don't necessarily run
+  // NewContext, so when InitializeContext is called this might never
+  // otherwise be called
+  InitializeContextRuntime(context);
+
   return true;
 }
 
