From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samuel Attard <samuel.r.attard@gmail.com>
Date: Wed, 14 Apr 2021 12:04:18 -0700
Subject: fix: the --harmony-weak-refs has been removed, remove from specs

Refs: https://chromium-review.googlesource.com/c/v8/v8/+/2741582

diff --git a/test/parallel/test-weakref.js b/test/parallel/test-weakref.js
index 9dac8463760dac6ee9270fe6805699c6bffd7299..ca7485aaa6a40c916fd3fec80960e9c6f6f3eda6 100644
--- a/test/parallel/test-weakref.js
+++ b/test/parallel/test-weakref.js
@@ -1,6 +1,6 @@
 'use strict';
 
-// Flags: --expose-gc --harmony-weak-refs
+// Flags: --expose-gc
 
 require('../common');
 const assert = require('assert');
