From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samuel Attard <sattard@slack-corp.com>
Date: Mon, 15 Jul 2019 14:20:10 -0700
Subject: fix: extern the NativeModuleEnv and options parser for debug builds


diff --git a/src/node_native_module_env.h b/src/node_native_module_env.h
index f662c67be50d404ee5b6cf6e2b8dd5991c59e723..da2d458344fb1006ac11c5b19c6ac5fbbca66a08 100644
--- a/src/node_native_module_env.h
+++ b/src/node_native_module_env.h
@@ -12,7 +12,7 @@ namespace native_module {
 
 extern const bool has_code_cache;
 
-class NativeModuleEnv {
+class NODE_EXTERN NativeModuleEnv {
  public:
   static void Initialize(v8::Local<v8::Object> target,
                          v8::Local<v8::Value> unused,
diff --git a/src/node_options.cc b/src/node_options.cc
index b9db121250c77e1bc8c35af336361ee444c271e4..a8e287279b31cba12d18300d08f0fd3856069a03 100644
--- a/src/node_options.cc
+++ b/src/node_options.cc
@@ -243,7 +243,7 @@ const EnvironmentOptionsParser _eop_instance{_dop_instance};
 
 // This Parse is not dead code. It is used by embedders (e.g., Electron).
 template <>
-void Parse(
+void NODE_EXTERN Parse(
   StringVector* const args, StringVector* const exec_args,
   StringVector* const v8_args,
   DebugOptions* const options,
diff --git a/src/node_options.h b/src/node_options.h
index 4ce5551284bb5b1b4194905a9fe619f852933405..07f18b5a91c7c42da9f780408e077235a4985f57 100644
--- a/src/node_options.h
+++ b/src/node_options.h
@@ -44,7 +44,7 @@ class HostPort {
   int port_;
 };
 
-class Options {
+class NODE_EXTERN Options {
  public:
   virtual void CheckOptions(std::vector<std::string>* errors) {}
   virtual ~Options() = default;
@@ -444,7 +444,7 @@ class NODE_EXTERN DebugOptionsParser : public OptionsParser<DebugOptions> {
 
 using StringVector = std::vector<std::string>;
 template <class OptionsType, class = Options>
-void Parse(
+void NODE_EXTERN Parse(
   StringVector* const args, StringVector* const exec_args,
   StringVector* const v8_args, OptionsType* const options,
   OptionEnvvarSettings required_env_settings, StringVector* const errors);
