From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: John Kleinschmidt <jkleinsc@electronjs.org>
Date: Wed, 26 Oct 2022 16:14:20 -0400
Subject: Remove AllocationSpace::MAP_SPACE enum constant

This was removed in:
https://chromium-review.googlesource.com/c/v8/v8/+/3967841

diff --git a/test/parallel/test-v8-stats.js b/test/parallel/test-v8-stats.js
index 3af7ea1e4a4598dc4125ff78e426d6dc6a025c66..83b375bd3c5b5dbd5189d48ad560580883ac91f6 100644
--- a/test/parallel/test-v8-stats.js
+++ b/test/parallel/test-v8-stats.js
@@ -42,7 +42,6 @@ const expectedHeapSpaces = [
   'code_large_object_space',
   'code_space',
   'large_object_space',
-  'map_space',
   'new_large_object_space',
   'new_space',
   'old_space',
