From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: John Kleinschmidt <jkleinsc@electronjs.org>
Date: Thu, 26 May 2022 17:08:33 -0400
Subject: JSON.parse errors made user-friendly

Update tests for  https://chromium-review.googlesource.com/c/v8/v8/+/3513684

diff --git a/test/es-module/test-esm-data-urls.js b/test/es-module/test-esm-data-urls.js
index 5be45d0f7af3b6dd483fe3b185c76e41a1bf533b..c20dd4f9067203bf40d35c92c2ef1bb27ec14b99 100644
--- a/test/es-module/test-esm-data-urls.js
+++ b/test/es-module/test-esm-data-urls.js
@@ -76,7 +76,7 @@ function createBase64URL(mime, body) {
       import('data:application/json;foo="test,",0',
         { assert: { type: 'json' } }), {
       name: 'SyntaxError',
-      message: /Unexpected end of JSON input/
+      message: 'data:application/json;foo="test,",0: Unterminated string in JSON at position 3'
     });
   }
   {
diff --git a/test/es-module/test-esm-invalid-pjson.js b/test/es-module/test-esm-invalid-pjson.js
index f3a38018637aa349ad79617ab9835e61d7058fe9..bc78b870c84c4baedecdd7ffc1157c86c307cebf 100644
--- a/test/es-module/test-esm-invalid-pjson.js
+++ b/test/es-module/test-esm-invalid-pjson.js
@@ -18,7 +18,7 @@ describe('ESM: Package.json', { concurrency: true }, () => {
       stderr.includes(
         `[ERR_INVALID_PACKAGE_CONFIG]: Invalid package config ${invalidJson} ` +
         `while importing "invalid-pjson" from ${entry}. ` +
-        `Unexpected token } in JSON at position ${12 + checkoutEOL.length * 2}`
+        `Expected ':' after property name in JSON at position ${12 + checkoutEOL.length * 2}`
       ),
       stderr
     );
