From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: John Kleinschmidt <jkleinsc@electronjs.org>
Date: Fri, 7 Nov 2025 16:36:35 -0500
Subject: Reland "[temporal] Unflag Temporal"

https://chromium-review.googlesource.com/c/v8/v8/+/7123876

diff --git a/test/common/globals.js b/test/common/globals.js
index 1641c4df36b92dbe29b06817d87130a94ae1b22f..114cd17c3fc44d7de885aada2a84f6fe6c9bfc0e 100644
--- a/test/common/globals.js
+++ b/test/common/globals.js
@@ -68,6 +68,7 @@ const intrinsics = new Set([
   'SuppressedError',
   'DisposableStack',
   'AsyncDisposableStack',
+  'Temporal'
 ]);
 
 if (global.gc) {
