From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Shelley Vohr <shelley.vohr@gmail.com>
Date: Wed, 12 Oct 2022 21:28:57 +0200
Subject: src: iwyu in cleanup_queue.cc

Upstreamed in https://github.com/nodejs/node/pull/44983.

diff --git a/src/cleanup_queue.cc b/src/cleanup_queue.cc
index 5235513f16c2574a88675d5a2f19d9cae3c417ac..6290b6796c532702fb5e7549ff0c3ad14d2b89d8 100644
--- a/src/cleanup_queue.cc
+++ b/src/cleanup_queue.cc
@@ -1,4 +1,5 @@
 #include "cleanup_queue.h"  // NOLINT(build/include_inline)
+#include <algorithm>
 #include <vector>
 #include "cleanup_queue-inl.h"
 
