From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: David Benjamin <davidben@google.com>
Date: Sun Oct 14 11:01:40 2018 -0500
Subject: Add a compatibility EVP_CIPH_OCB_MODE value.


diff --git a/include/openssl/cipher.h b/include/openssl/cipher.h
index 5963413..e9545c8 100644
--- a/include/openssl/cipher.h
+++ b/include/openssl/cipher.h
@@ -424,8 +424,9 @@
 
 // The following flags do nothing and are included only to make it easier to
 // compile code with BoringSSL.
-#define EVP_CIPH_CCM_MODE 0
-#define EVP_CIPH_WRAP_MODE 0
+#define EVP_CIPH_CCM_MODE (-1)
+#define EVP_CIPH_OCB_MODE (-2)
+#define EVP_CIPH_WRAP_MODE (-3)
 #define EVP_CIPHER_CTX_FLAG_WRAP_ALLOW 0
 
 // EVP_CIPHER_CTX_set_flags does nothing.
