From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nitish Sakhawalkar <nitsakh@icloud.com>
Date: Thu, 7 Mar 2019 17:02:13 -0800
Subject: src: use CHECK(false) in switch default case

Porting https://github.com/nodejs/node/pull/26502

diff --git a/src/node_crypto.cc b/src/node_crypto.cc
index c56e488c30784739cf436a0029bdead15aed7541..951dfc447c25762b6f43ed1b86bc4c75ec423dc7 100644
--- a/src/node_crypto.cc
+++ b/src/node_crypto.cc
@@ -3447,7 +3447,7 @@ static ManagedEVPPKey GetPublicOrPrivateKeyFromJs(
           is_public = false;
           break;
         default:
-          CHECK(!"Invalid key encoding type");
+          CHECK(false);
       }
 
       if (is_public) {
