diff --git a/third_party/zlib/BUILD.gn b/third_party/zlib/BUILD.gn
index 902e287f8a0a..c08a187ca968 100644
--- a/third_party/zlib/BUILD.gn
+++ b/third_party/zlib/BUILD.gn
@@ -265,6 +265,10 @@ static_library("zlib") {
   defines = []
   deps = []

+  if (is_win) {
+    defines += [ "ZLIB_DLL" ]
+  }
+
   if (!is_ios && (current_cpu == "x86" || current_cpu == "x64")) {
     deps += [ ":zlib_crc32_simd" ]

