diff --git a/base/BUILD.gn b/base/BUILD.gn
index 943412faefd7..c505df950b2b 100644
--- a/base/BUILD.gn
+++ b/base/BUILD.gn
@@ -1217,7 +1217,7 @@ jumbo_component("base") {
   # Needed for <atomic> if using newer C++ library than sysroot, except if
   # building inside the cros_sdk environment - use host_toolchain as a
   # more robust check for this.
-  if (!use_sysroot && (is_android || (is_linux && !is_chromecast)) &&
+  if ((is_android || (is_linux && !is_chromecast)) &&
       host_toolchain != "//build/toolchain/cros:host") {
     libs = [ "atomic" ]
   }
