From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Cheng Zhao <zcbenz@gmail.com>
Date: Thu, 20 Sep 2018 17:49:42 -0700
Subject: allow_webview_file_url.patch

Allow webview to load non-web URLs.

diff --git a/content/browser/loader/resource_dispatcher_host_impl.cc b/content/browser/loader/resource_dispatcher_host_impl.cc
index 5d721cad1ed0d42441edd57c8c17b559abbbf0a7..c100b8b7d645a2683aee410234cc3d05335eb7ea 100644
--- a/content/browser/loader/resource_dispatcher_host_impl.cc
+++ b/content/browser/loader/resource_dispatcher_host_impl.cc
@@ -1473,6 +1473,8 @@ void ResourceDispatcherHostImpl::BeginNavigationRequest(
       !policy->IsWebSafeScheme(info.common_params.url.scheme()) &&
       !is_external_protocol;
 
+  non_web_url_in_guest = false;
+
   if (is_shutdown_ || non_web_url_in_guest) {
     url_loader_client->OnComplete(
         network::URLLoaderCompletionStatus(net::ERR_ABORTED));
