diff --git a/services/device/BUILD.gn b/services/device/BUILD.gn
index 20301dbd005d..31a7568a8aa8 100644
--- a/services/device/BUILD.gn
+++ b/services/device/BUILD.gn
@@ -18,6 +18,7 @@ source_set("lib") {
   # dependence should only be for the purpose of embedding the Device Service.
   visibility = [
     ":test_support",
+    "//libchromiumcontent/chromiumcontent:device_service",
     "//content/browser",
   ]
   sources = [
