diff --git a/ui/views/widget/widget.cc b/ui/views/widget/widget.cc
index bf4b1d07f7fe..8435c60a0786 100644
--- a/ui/views/widget/widget.cc
+++ b/ui/views/widget/widget.cc
@@ -995,7 +995,8 @@ bool Widget::IsTranslucentWindowOpacitySupported() const {
 
 void Widget::OnSizeConstraintsChanged() {
   native_widget_->OnSizeConstraintsChanged();
-  non_client_view_->SizeConstraintsChanged();
+  if (non_client_view_)
+    non_client_view_->SizeConstraintsChanged();
 }
 
 void Widget::OnOwnerClosing() {}
