From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tomas Rycl <torycl@microsoft.com>
Date: Thu, 20 Sep 2018 17:47:25 -0700
Subject: browser_plugin_guest.patch


diff --git a/content/browser/browser_plugin/browser_plugin_guest.cc b/content/browser/browser_plugin/browser_plugin_guest.cc
index e5f885ab4bc06672fb1f9ac7f76bd92203894559..d46eee5fcffe2137bdc31bf4a77691af1e091143 100644
--- a/content/browser/browser_plugin/browser_plugin_guest.cc
+++ b/content/browser/browser_plugin/browser_plugin_guest.cc
@@ -207,8 +207,11 @@ void BrowserPluginGuest::Init() {
 
   WebContentsImpl* owner_web_contents = static_cast<WebContentsImpl*>(
       delegate_->GetOwnerWebContents());
-  owner_web_contents->CreateBrowserPluginEmbedderIfNecessary();
-  InitInternal(BrowserPluginHostMsg_Attach_Params(), owner_web_contents);
+  if (nullptr != owner_web_contents)
+  {
+      owner_web_contents->CreateBrowserPluginEmbedderIfNecessary();
+      InitInternal(BrowserPluginHostMsg_Attach_Params(), owner_web_contents);
+  }
 }
 
 base::WeakPtr<BrowserPluginGuest> BrowserPluginGuest::AsWeakPtr() {
