From 790e1e19af133252e95e3a4de8c8e7a3a011bf02 Mon Sep 17 00:00:00 2001
From: deepak1556 <hop2deep@gmail.com>
Date: Thu, 20 Sep 2018 17:50:48 -0700
Subject: disable_extensions_gn.patch

Fix build files generation when chrome extensions are disabled.

diff --git a/chrome/browser/apps/app_shim/BUILD.gn b/chrome/browser/apps/app_shim/BUILD.gn
index f8a6d1868788..350c3572ec54 100644
--- a/chrome/browser/apps/app_shim/BUILD.gn
+++ b/chrome/browser/apps/app_shim/BUILD.gn
@@ -1,6 +1,7 @@
 # Copyright 2015 The Chromium Authors. All rights reserved.
 # Use of this source code is governed by a BSD-style license that can be
 # found in the LICENSE file.
+import("//extensions/buildflags/buildflags.gni")
 
 # This is the part of the Chrome browser process responsible for launching and
 # communicating with app_shim processes on Mac.
@@ -14,8 +15,6 @@ source_set("app_shim") {
     "app_shim_host_manager_mac.mm",
     "apps_page_shim_handler.h",
     "apps_page_shim_handler.mm",
-    "extension_app_shim_handler_mac.cc",
-    "extension_app_shim_handler_mac.h",
     "unix_domain_socket_acceptor.cc",
     "unix_domain_socket_acceptor.h",
   ]
@@ -24,7 +23,16 @@ source_set("app_shim") {
     "//chrome/common:mojo_bindings",
     "//content/public/browser",
     "//content/public/common",
-    "//extensions/browser",
-    "//extensions/common",
   ]
+
+  if (enable_extensions) {
+    sources += [
+      "extension_app_shim_handler_mac.cc",
+      "extension_app_shim_handler_mac.h",
+    ]
+    deps += [
+      "//extensions/browser",
+      "//extensions/common",
+    ]
+  }
 }
diff --git a/chrome/browser/ui/BUILD.gn b/chrome/browser/ui/BUILD.gn
index 1b14d628a63b..00401612c191 100644
--- a/chrome/browser/ui/BUILD.gn
+++ b/chrome/browser/ui/BUILD.gn
@@ -2570,7 +2570,10 @@ split_static_library("ui") {
         "views/tabs/window_finder_mac.mm",
       ]
 
-      deps += [ "//extensions/components/native_app_window" ]
+
+      if (enable_extensions) {
+        deps += [ "//extensions/components/native_app_window" ]
+      }
 
       # Truly cocoa-browser-specific sources. These are secondary UI pieces that
       # are obsolete before mac_views_browser will ever ship, so they aren't
-- 
2.17.0
