From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samuel Attard <sattard@slack-corp.com>
Date: Fri, 26 Apr 2019 13:19:34 -0700
Subject: disable custom libcxx on windows


diff --git a/build/config/c++/c++.gni b/build/config/c++/c++.gni
index 678019a257e0bd9d43a659b1303fd99bae161849..3bf8e926d1e818ed77eb4b6119e37eef26b8b0ab 100644
--- a/build/config/c++/c++.gni
+++ b/build/config/c++/c++.gni
@@ -11,7 +11,7 @@ declare_args() {
   # Don't check in changes that set this to false for more platforms; doing so
   # is not supported.
   use_custom_libcxx =
-      is_fuchsia || is_android || is_mac || is_ios || (is_win && is_clang) ||
+      is_fuchsia || is_android || is_mac || is_ios || (is_win && is_clang && !is_electron_build) ||
       (is_linux &&
        (!is_chromeos || default_toolchain != "//build/toolchain/cros:target"))
 
