From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Cheng Zhao <zcbenz@gmail.com>
Date: Tue, 15 Jan 2019 14:57:02 -0700
Subject: disable_time_ticks_dcheck.patch

The DCHECK is failing for some reason.

diff --git a/content/common/inter_process_time_ticks_converter.cc b/content/common/inter_process_time_ticks_converter.cc
index 128abab37eb8a96535ef92ebf11a463e863cc485..4d8e5e9c05b11083a69537d5badc85924b6fbae2 100644
--- a/content/common/inter_process_time_ticks_converter.cc
+++ b/content/common/inter_process_time_ticks_converter.cc
@@ -55,13 +55,13 @@ LocalTimeTicks InterProcessTimeTicksConverter::ToLocalTimeTicks(
 
   RemoteTimeDelta remote_delta = remote_time_ticks - remote_lower_bound_;
 
-  DCHECK_LE(remote_time_ticks, remote_upper_bound_);
+  // DCHECK_LE(remote_time_ticks, remote_upper_bound_);
   return local_base_time_ + ToLocalTimeDelta(remote_delta);
 }
 
 LocalTimeDelta InterProcessTimeTicksConverter::ToLocalTimeDelta(
     RemoteTimeDelta remote_delta) const {
-  DCHECK_LE(remote_lower_bound_ + remote_delta, remote_upper_bound_);
+  // DCHECK_LE(remote_lower_bound_ + remote_delta, remote_upper_bound_);
 
   // For remote times that come before remote time range, apply just time
   // offset and ignore scaling, so as to avoid extrapolation error for values
