From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samuel Attard <sattard@slack-corp.com>
Date: Fri, 29 Mar 2019 11:04:46 -0700
Subject: disable custom libcxx on windows


diff --git a/build/config/c++/c++.gni b/build/config/c++/c++.gni
index 5d45ff18455275e2ab416030fb01b103c7a630cf..39774eb51b112e02a9aa9c5288e8b52e33fdcb17 100644
--- a/build/config/c++/c++.gni
+++ b/build/config/c++/c++.gni
@@ -13,7 +13,8 @@ declare_args() {
   # TODO(https://crbug.com/942939): Re-enable libc++ on Windows with libfuzzer.
   use_custom_libcxx =
       is_fuchsia || is_android || is_mac ||
-      (is_win && is_clang && !use_libfuzzer) ||
+      # Do not use custom libcxx on windows
+      # (is_win && is_clang && !use_libfuzzer) ||
       (is_linux &&
        (!is_chromeos || default_toolchain != "//build/toolchain/cros:target"))
 
