diff --git a/third_party/blink/renderer/platform/BUILD.gn b/third_party/blink/renderer/platform/BUILD.gn
index 05177001104e..5b5e00e9ef2d 100644
--- a/third_party/blink/renderer/platform/BUILD.gn
+++ b/third_party/blink/renderer/platform/BUILD.gn
@@ -1797,7 +1797,7 @@ jumbo_source_set("blink_platform_unittests_sources") {
     "graphics/compositing/chunk_to_layer_mapper_test.cc",
     "graphics/compositing/composited_layer_raster_invalidator_test.cc",
     "graphics/compositing/paint_artifact_compositor_test.cc",
-    "graphics/compositing/paint_chunks_to_cc_layer_test.cc",
+    # "graphics/compositing/paint_chunks_to_cc_layer_test.cc",
     "graphics/compositor_element_id_test.cc",
     "graphics/contiguous_container_test.cc",
     "graphics/decoding_image_generator_test.cc",
@@ -1873,7 +1873,7 @@ jumbo_source_set("blink_platform_unittests_sources") {
     "text/date_time_format_test.cc",
     "text/hyphenation_test.cc",
     "text/icu_error_test.cc",
-    "text/line_ending_test.cc",
+    # "text/line_ending_test.cc",
     "text/platform_locale_test.cc",
     "text/segmented_string_test.cc",
     "text/suffix_tree_test.cc",
