From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jeremy Apthorp <nornagon@nornagon.net>
Date: Thu, 21 Feb 2019 17:17:51 -0800
Subject: FIXME: disable field-trial descriptor DCHECK


diff --git a/content/browser/child_process_launcher_helper_posix.cc b/content/browser/child_process_launcher_helper_posix.cc
index 30f424d662bdc72e692bf86d0270ef1cb50fd8fc..2b2ed424152f5feccad81eef3c89892d6189f477 100644
--- a/content/browser/child_process_launcher_helper_posix.cc
+++ b/content/browser/child_process_launcher_helper_posix.cc
@@ -83,8 +83,9 @@ std::unique_ptr<PosixFileDescriptorInfo> CreateDefaultPosixFilesToMap(
 // Mac shared memory doesn't use file descriptors.
 #if !defined(OS_MACOSX)
   int fd = base::FieldTrialList::GetFieldTrialDescriptor();
-  DCHECK_NE(fd, -1);
-  files_to_register->Share(service_manager::kFieldTrialDescriptor, fd);
+  //DCHECK_NE(fd, -1);
+  if (fd >= 0)
+    files_to_register->Share(service_manager::kFieldTrialDescriptor, fd);
 #endif
 
   DCHECK(mojo_channel_remote_endpoint.is_valid());
