From c7042021cfd9d4e50b107e372225a14f7492bce5 Mon Sep 17 00:00:00 2001
From: Jeremy Apthorp <nornagon@nornagon.net>
Date: Thu, 20 Sep 2018 17:49:34 -0700
Subject: mas-lssetapplicationlaunchservicesserverconnectionstatus.patch

Removes usage of the _LSSetApplicationLaunchServicesServerConnectionStatus
private API.

diff --git a/content/gpu/gpu_main.cc b/content/gpu/gpu_main.cc
index f14c045d47ad..9c8751a6fe0b 100644
--- a/content/gpu/gpu_main.cc
+++ b/content/gpu/gpu_main.cc
@@ -274,8 +274,10 @@ int GpuMain(const MainFunctionParams& parameters) {
     std::unique_ptr<base::MessagePump> pump(new base::MessagePumpNSRunLoop());
     main_message_loop.reset(new base::MessageLoop(std::move(pump)));
 
+#ifndef MAS_BUILD
     // Tell LaunchServices to continue without a connection to the daemon.
     _LSSetApplicationLaunchServicesServerConnectionStatus(0, nullptr);
+#endif
 #else
     main_message_loop.reset(
         new base::MessageLoop(base::MessageLoop::TYPE_DEFAULT));
-- 
2.17.0

