From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jeremy Apthorp <nornagon@nornagon.net>
Date: Thu, 20 Sep 2018 17:49:34 -0700
Subject: mas-lssetapplicationlaunchservicesserverconnectionstatus.patch

Removes usage of the _LSSetApplicationLaunchServicesServerConnectionStatus
private API.

diff --git a/content/gpu/gpu_main.cc b/content/gpu/gpu_main.cc
index 0b9d501e7b0da03c2bc668944d6b4172c71dae72..0ad797171bd359037bcf8d3e7c9e68d3408cd49c 100644
--- a/content/gpu/gpu_main.cc
+++ b/content/gpu/gpu_main.cc
@@ -280,8 +280,10 @@ int GpuMain(const MainFunctionParams& parameters) {
     std::unique_ptr<base::MessagePump> pump(new base::MessagePumpNSRunLoop());
     main_message_loop.reset(new base::MessageLoop(std::move(pump)));
 
+#ifndef MAS_BUILD
     // Tell LaunchServices to continue without a connection to the daemon.
     _LSSetApplicationLaunchServicesServerConnectionStatus(0, nullptr);
+#endif
 #else
     main_message_loop.reset(
         new base::MessageLoop(base::MessageLoop::TYPE_DEFAULT));
