From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aleksei Kuzmin <alkuzmin@microsoft.com>
Date: Thu, 20 Sep 2018 17:46:10 -0700
Subject: no_stack_dumping.patch


diff --git a/content/app/content_main_runner_impl.cc b/content/app/content_main_runner_impl.cc
index 61ef09f69587d68109db397aca17b2a0f89a8af6..1fecc396990b0cca40bf0c42ce107f6c4e5a01c5 100644
--- a/content/app/content_main_runner_impl.cc
+++ b/content/app/content_main_runner_impl.cc
@@ -776,7 +776,7 @@ int ContentMainRunnerImpl::Initialize(const ContentMainParams& params) {
           return nullptr;
         }));
 
-#if !defined(OFFICIAL_BUILD)
+#if 0
 #if defined(OS_WIN)
     bool should_enable_stack_dump = !process_type.empty();
 #else
