diff --git a/services/device/wake_lock/power_save_blocker/BUILD.gn b/services/device/wake_lock/power_save_blocker/BUILD.gn
index dfce90f6b791..a2f7f776f4d8 100644
--- a/services/device/wake_lock/power_save_blocker/BUILD.gn
+++ b/services/device/wake_lock/power_save_blocker/BUILD.gn
@@ -9,7 +9,7 @@ if (is_android) {
   import("//build/config/android/rules.gni")
 }
 
-source_set("power_save_blocker") {
+static_library("power_save_blocker") {
   visibility = [
     # //remoting runs in a separate process which is outside of the context of
     # the ServiceManager-based world. Instead of embedding a Service Manager
@@ -18,6 +18,7 @@ source_set("power_save_blocker") {
     "//remoting/host:*",
     "//remoting/host/win:*",
     "//services/device/wake_lock:*",
+    "//electron:*",
   ]
 
   sources = [
