diff --git a/content/browser/loader/stream_resource_handler.h b/content/browser/loader/stream_resource_handler.h
index cb8a2d58d6b8..dbfa64089325 100644
--- a/content/browser/loader/stream_resource_handler.h
+++ b/content/browser/loader/stream_resource_handler.h
@@ -11,6 +11,7 @@
 #include "base/memory/ref_counted.h"
 #include "content/browser/loader/resource_handler.h"
 #include "content/browser/loader/stream_writer.h"
+#include "content/common/content_export.h"
 
 namespace net {
 class URLRequest;
@@ -22,7 +23,7 @@ class ResourceController;
 class StreamRegistry;
 
 // Redirect this resource to a stream.
-class StreamResourceHandler : public ResourceHandler {
+class CONTENT_EXPORT StreamResourceHandler : public ResourceHandler {
  public:
   // |origin| will be used to construct the URL for the Stream. See
   // WebCore::BlobURL and and WebCore::SecurityOrigin in Blink to understand
