From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: deepak1556 <hop2deep@gmail.com>
Date: Thu, 20 Sep 2018 17:46:39 -0700
Subject: stream_resource_handler.patch


diff --git a/content/browser/loader/stream_resource_handler.h b/content/browser/loader/stream_resource_handler.h
index b6f51bba48cd145d6bccaff649d7875cf0520b7d..1ee88dfacee3e9bf3e4393a32ebd719251709fb3 100644
--- a/content/browser/loader/stream_resource_handler.h
+++ b/content/browser/loader/stream_resource_handler.h
@@ -11,6 +11,7 @@
 #include "base/memory/ref_counted.h"
 #include "content/browser/loader/resource_handler.h"
 #include "content/browser/loader/stream_writer.h"
+#include "content/common/content_export.h"
 
 namespace net {
 class URLRequest;
@@ -22,7 +23,7 @@ class ResourceController;
 class StreamRegistry;
 
 // Redirect this resource to a stream.
-class StreamResourceHandler : public ResourceHandler {
+class CONTENT_EXPORT StreamResourceHandler : public ResourceHandler {
  public:
   // |origin| will be used to construct the URL for the Stream. See
   // WebCore::BlobURL and and WebCore::SecurityOrigin in Blink to understand
