diff --git a/chrome/browser/speech/tts_controller_impl.cc b/chrome/browser/speech/tts_controller_impl.cc
index 2ca56c3a247e..f58e33b3c019 100644
--- a/chrome/browser/speech/tts_controller_impl.cc
+++ b/chrome/browser/speech/tts_controller_impl.cc
@@ -624,12 +624,14 @@ const PrefService* TtsControllerImpl::GetPrefService(
     const Utterance* utterance) {
   const PrefService* prefs = nullptr;
   // The utterance->browser_context() is null in tests.
+#if 0
   if (utterance->browser_context()) {
     const Profile* profile =
         Profile::FromBrowserContext(utterance->browser_context());
     if (profile)
       prefs = profile->GetPrefs();
   }
+#endif
   return prefs;
 }

diff --git a/chrome/browser/speech/tts_message_filter.cc b/chrome/browser/speech/tts_message_filter.cc
index 013c7a9c60f9..7f468ab165f9 100644
--- a/chrome/browser/speech/tts_message_filter.cc
+++ b/chrome/browser/speech/tts_message_filter.cc
@@ -24,12 +24,14 @@ TtsMessageFilter::TtsMessageFilter(content::BrowserContext* browser_context)
   CHECK(BrowserThread::CurrentlyOn(BrowserThread::UI));
   TtsController::GetInstance()->AddVoicesChangedDelegate(this);

+#if 0
   // TODO(dmazzoni): make it so that we can listen for a BrowserContext
   // being destroyed rather than a Profile.  http://crbug.com/444668
   Profile* profile = Profile::FromBrowserContext(browser_context);
   notification_registrar_.Add(this,
                               chrome::NOTIFICATION_PROFILE_DESTROYED,
                               content::Source<Profile>(profile));
+#endif

   // Balanced in OnChannelClosingInUIThread() to keep the ref-count be non-zero
   // until all pointers to this class are invalidated.
