From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jeremy Apthorp <nornagon@nornagon.net>
Date: Tue, 16 Apr 2019 11:25:08 -0700
Subject: unsandboxed ppapi processes skip zygote


diff --git a/content/browser/ppapi_plugin_process_host.cc b/content/browser/ppapi_plugin_process_host.cc
index d7d70e9ad63257e77b109121334311a9930725bf..d3d7f18bb2d3a51ed8b3ae2d83b1e0e33838b444 100644
--- a/content/browser/ppapi_plugin_process_host.cc
+++ b/content/browser/ppapi_plugin_process_host.cc
@@ -105,6 +105,9 @@ class PpapiPluginSandboxedProcessLauncherDelegate
   service_manager::ZygoteHandle GetZygote() override {
     const base::CommandLine& browser_command_line =
         *base::CommandLine::ForCurrentProcess();
+    if (browser_command_line.HasSwitch(service_manager::switches::kNoSandbox)) {
+      return nullptr;
+    }
     base::CommandLine::StringType plugin_launcher = browser_command_line
         .GetSwitchValueNative(switches::kPpapiPluginLauncher);
     if (is_broker_ || !plugin_launcher.empty())
